{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='contents'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Figures and analysis\n",
    "\n",
    "This notebook contains scripts for producing the main figures and results accompanying the manuscript. Here we perform basic organization and processing of data, which is then passed to functions in `figures.py` and `mplot.py` (available at this [GitHub repository](https://github.com/johnbarton/mplot)) for detailed formatting. The figures produced are stored as PDFs in the `/figures` folder.\n",
    "\n",
    "## Contents\n",
    "\n",
    "- [Overview and table of contents](#contents)\n",
    "- [Loading libraries and global variables](#global)\n",
    "- [Figures and data analysis](#figures)  \n",
    "    - [Figure 1](#fig1-figs1)  \n",
    "    - [Figure 2](#fig2)\n",
    "    - [Figure 3](#fig3)\n",
    "    - [Figure 4](#fig4-figs9)\n",
    "    - [Supplementary Figure 1](#fig1-figs1)\n",
    "    - [Supplementary Figure 2](#figs2)\n",
    "    - [Supplementary Figure 3](#figs3-figs6)\n",
    "    - [Supplementary Figure 4](#figs4)\n",
    "    - [Supplementary Figure 5](#figs5)\n",
    "    - [Supplementary Figure 6](#figs3-figs6)\n",
    "    - [Supplementary Figure 7](#figs7-figs8)\n",
    "    - [Supplementary Figure 8](#figs7-figs8)\n",
    "    - [Supplementary Figure 9](#fig4-figs9)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='global'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Libraries and variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This notebook was prepared using:\n",
      "python version 3.7.4 (default, Jul  9 2019, 18:13:23) \n",
      "[Clang 10.0.1 (clang-1001.0.46.4)]\n",
      "numpy version 1.16.4\n",
      "scipy version 1.3.0\n",
      "pandas version 0.24.2\n",
      "matplotlib version 3.1.1\n"
     ]
    }
   ],
   "source": [
    "# Full library list and version numbers\n",
    "\n",
    "print('This notebook was prepared using:')\n",
    "\n",
    "import sys, os\n",
    "from copy import deepcopy\n",
    "from importlib import reload\n",
    "print('python version %s' % sys.version)\n",
    "\n",
    "import numpy as np\n",
    "print('numpy version %s' % np.__version__)\n",
    "\n",
    "import scipy as sp\n",
    "import scipy.stats as st\n",
    "print('scipy version %s' % sp.__version__)\n",
    "\n",
    "import pandas as pd\n",
    "print('pandas version %s' % pd.__version__)\n",
    "\n",
    "import matplotlib\n",
    "import matplotlib.cm as cm\n",
    "import matplotlib.pyplot as plot\n",
    "import matplotlib.gridspec as gridspec\n",
    "import matplotlib.image as mpimg\n",
    "print('matplotlib version %s' % matplotlib.__version__)\n",
    "\n",
    "import figures as fig\n",
    "import mplot as mp\n",
    "\n",
    "\n",
    "# GLOBAL VARIABLES\n",
    "\n",
    "NUC = ['-', 'A', 'C', 'G', 'T']\n",
    "REF = NUC[0]\n",
    "CONS_TAG = 'CONSENSUS'\n",
    "HXB2_TAG = 'B.FR.1983.HXB2-LAI-IIIB-BRU.K03455.19535'\n",
    "\n",
    "MPL_DIR = 'src/MPL'\n",
    "HIV_DIR = 'data/HIV'\n",
    "SIM_DIR = 'data/simulation'\n",
    "\n",
    "TESTS   = [   'example',      'medium_simple',      'medium_complex']\n",
    "N_VALS  = dict(example=  1000, medium_simple=  1000, medium_complex=1000)\n",
    "L_VALS  = dict(example=    50, medium_simple=    50, medium_complex=  50)\n",
    "T0_VALS = dict(example=     0, medium_simple=     0, medium_complex=  10)\n",
    "T_VALS  = dict(example=   400, medium_simple=  1000, medium_complex= 310)\n",
    "MU_VALS = dict(example=  1e-3, medium_simple=  1e-4, medium_complex=1e-4)\n",
    "NB_VALS = dict(example=    10, medium_simple=    10, medium_complex=  10)\n",
    "ND_VALS = dict(example=    10, medium_simple=    10, medium_complex=  10)\n",
    "SB_VALS = dict(example= 0.025, medium_simple= 0.025, medium_complex= 0.1)\n",
    "SD_VALS = dict(example=-0.025, medium_simple=-0.025, medium_complex=-0.1)\n",
    "\n",
    "N_TRIALS     =  100  # number of independent trials to run for each test set\n",
    "COMP_NS_VALS = [100] # number of sequence samples to collect per time point \n",
    "COMP_DT_VALS = [ 10] # time between sampling events (in discrete generations)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='figures'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Figures and data analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='fig1-figs1'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Figure 1. Example evolutionary path for a 50-site system and inferred selection coefficients.  \n",
    "### Supplementary Figure 1. Plot of paths for individual alleles, inferred selection coefficients, and aggregate properties for different levels of sampling."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Figure 1 done.\n",
      "Figure S1 done.\n",
      "\n",
      "example AUROC (beneficial, deleterious): (0.99, 0.99)\n",
      "example NRMSE: 0.45\n",
      "\n",
      "medium_complex AUROC (beneficial, deleterious): (0.92, 0.93)\n",
      "medium_complex NRMSE: 0.75\n",
      "\n",
      "medium_complex LLS best AUROC beneficial:  0.83\n",
      "medium_complex FIT best AUROC deleterious: 0.81\n",
      "medium_complex CLEAR best NRMSE: 0.90\n"
     ]
    }
   ],
   "source": [
    "from imp import reload\n",
    "reload(mp)\n",
    "reload(fig)\n",
    "\n",
    "\n",
    "pdata = {\n",
    "    'n_gen':  400,                               # number of generations\n",
    "    'dg':     5,                                 # spacing between generations for plot\n",
    "    'N':      1000,                              # population size\n",
    "    'xfile':  'wfsim_example_0_T400_ns1000_dt1', # file in directory data/ containing example trajectory\n",
    "    'method': 'MPL',                             # inference method\n",
    "    'n_ben':  10,                                # number of beneficial mutations\n",
    "    'n_neu':  30,                                # number of neutral mutations\n",
    "    'n_del':  10,                                # number of deleterious mutations\n",
    "    's_ben':  0.025,                             # selection coefficient of beneficial mutations\n",
    "    's_neu':  0,                                 # selection coefficient of neutral mutations\n",
    "    's_del':  -0.025,                            # selection coefficient of deleterious mutations\n",
    "    'r_seed': 1,                                 # random seed for scattering inferred selection coefficients\n",
    "}\n",
    "\n",
    "fig.plot_figure_example_mpl(**pdata)\n",
    "fig.plot_supplementary_figure_example_mpl(**pdata)\n",
    "\n",
    "test = 'example'\n",
    "df   = pd.read_csv('%s/MPL_%s_collected_extended.csv.gz' % (SIM_DIR, test), memory_map=True)\n",
    "df   = df[df.method=='MPL']\n",
    "df_s = df[(df.deltat==1) & (df.ns==1000)]\n",
    "\n",
    "print('')\n",
    "print('%s AUROC (beneficial, deleterious): (%.2f, %.2f)' % (test, np.mean(df_s.AUROC_ben), np.mean(df_s.AUROC_del)))\n",
    "\n",
    "dscols = [j for j in df_s.columns if 'ds' in j]\n",
    "scols  = [j for j in df_s.columns if ('s' in j) and (j not in dscols+['ns'])]\n",
    "y_num = np.sqrt(np.sum(np.array(df_s[dscols]**2),axis=1))\n",
    "y_den = np.sqrt(np.sum([NB_VALS[test] * (SB_VALS[test]**2), ND_VALS[test] * (SD_VALS[test]**2)]))\n",
    "NRMSE = np.mean(np.array(y_num)/np.array(y_den))\n",
    "\n",
    "print('%s NRMSE: %.2f' % (test, NRMSE))\n",
    "\n",
    "\n",
    "test = 'medium_complex'\n",
    "df   = pd.read_csv('%s/MPL_%s_collected_extended.csv.gz' % (SIM_DIR, test), memory_map=True)\n",
    "df   = df[df.method=='MPL']\n",
    "df_s = df[(df.deltat==10) & (df.ns==100)]\n",
    "\n",
    "print('')\n",
    "print('%s AUROC (beneficial, deleterious): (%.2f, %.2f)' % (test, np.mean(df_s.AUROC_ben), np.mean(df_s.AUROC_del)))\n",
    "\n",
    "dscols = [j for j in df_s.columns if 'ds' in j]\n",
    "scols  = [j for j in df_s.columns if ('s' in j) and (j not in dscols+['ns'])]\n",
    "y_num = np.sqrt(np.sum(np.array(df_s[dscols]**2),axis=1))\n",
    "y_den = np.sqrt(np.sum([NB_VALS[test] * (SB_VALS[test]**2), ND_VALS[test] * (SD_VALS[test]**2)]))\n",
    "NRMSE = np.mean(np.array(y_num)/np.array(y_den))\n",
    "\n",
    "print('%s NRMSE: %.2f' % (test, NRMSE))\n",
    "\n",
    "\n",
    "methods = ['ApproxWF', 'CLEAR', 'EandR', 'FIT', 'IM', 'LLS', 'WFABC']\n",
    "best_NRMSE,     best_NRMSE_method     = 100, 'none'\n",
    "best_AUROC_ben, best_AUROC_ben_method =   0, 'none'\n",
    "best_AUROC_del, best_AUROC_del_method =   0, 'none'\n",
    "for m in methods:\n",
    "    df    = pd.read_csv('%s/%s_%s_collected_extended.csv.gz' % (SIM_DIR, m, test), memory_map=True)\n",
    "    y_num = np.sqrt(np.sum(np.array(df[dscols]**2),axis=1))\n",
    "    y_den = np.sqrt(np.sum([NB_VALS[test] * (SB_VALS[test]**2), ND_VALS[test] * (SD_VALS[test]**2)]))\n",
    "    NRMSE = np.mean(np.array(y_num)/np.array(y_den))\n",
    "    AUROC_ben = np.mean(df.AUROC_ben)\n",
    "    AUROC_del = np.mean(df.AUROC_del)\n",
    "    if NRMSE<best_NRMSE:\n",
    "        best_NRMSE = NRMSE\n",
    "        best_NRMSE_method = m\n",
    "    if AUROC_ben>best_AUROC_ben:\n",
    "        best_AUROC_ben = AUROC_ben\n",
    "        best_AUROC_ben_method = m\n",
    "    if AUROC_del>best_AUROC_del:\n",
    "        best_AUROC_del = AUROC_del\n",
    "        best_AUROC_del_method = m\n",
    "        \n",
    "print('')\n",
    "print('%s %s best AUROC beneficial:  %.2f' % (test, best_AUROC_ben_method, best_AUROC_ben))\n",
    "print('%s %s best AUROC deleterious: %.2f' % (test, best_AUROC_del_method, best_AUROC_del))\n",
    "print('%s %s best NRMSE: %.2f'             % (test, best_NRMSE_method, best_NRMSE))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='fig2'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Figure 2. Summary of selection on HIV during intrahost evolution.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MPL enrichment in escape mutations that are reversions (top 1%): 326\n",
      "(MPL, SL) enrichment in escape mutations (top 1%): (19, 11)\n",
      "\tMPL fraction of SL: 1.76\n",
      "(MPL, SL) enrichment in non-epitope reversions (top 1%): (14, 33)\n",
      "\tMPL fraction of SL: 0.42\n",
      "MPL percentage gain in escape mutations (top 1%): 76\n",
      "Figure 2 done.\n"
     ]
    }
   ],
   "source": [
    "from imp import reload\n",
    "reload(mp)\n",
    "reload(fig)\n",
    "\n",
    "\n",
    "# Load in selection data and do the analysis\n",
    "\n",
    "df = pd.read_csv('%s/analysis/total-selection.csv' % (HIV_DIR), comment='#', memory_map=True)\n",
    "df = df[df.ppt!='cap256']\n",
    "\n",
    "\n",
    "# Get fraction of top (1)2% most beneficial mutations that belong in different categories\n",
    "\n",
    "top          = int(np.round(len(df)*0.01))\n",
    "s_MPL_sorted = np.argsort(df.s_MPL)[::-1]\n",
    "s_SL_sorted  = np.argsort(df.s_SL)[::-1]\n",
    "s_set        = df.iloc[s_MPL_sorted[:top]]\n",
    "\n",
    "n_poly = []\n",
    "# Env exposed, non-epitope, nonsynonymous, no effect on glycosylation motifs\n",
    "n_poly.append(np.sum((s_set.nonsynonymous> 0) & (s_set.in_epitope==False) & (s_set.glycan==0)\n",
    "                      & (s_set.exposed==True))/top)\n",
    "# +/- glycosylation motif, non-epitope, nonsynonymous\n",
    "n_poly.append(np.sum((s_set.nonsynonymous> 0) & (s_set.in_epitope==False) & (s_set.glycan!=0))/top)\n",
    "# CD8+ T cell epitope, nonsynonymous\n",
    "n_poly.append(np.sum((s_set.nonsynonymous> 0) & (s_set.in_epitope==True))/top)\n",
    "# Flanking CD8+ T cell epitope, nonsynonymous, not Env exposed and no effect on glycosylation motifs\n",
    "n_poly.append(np.sum((s_set.nonsynonymous> 0) & (s_set.in_epitope==False) & (s_set.glycan==0)\n",
    "                      & (s_set.exposed==False) & (s_set.flanking>0))/top)\n",
    "# Synonymous, not reversion\n",
    "n_poly.append(np.sum((s_set.nonsynonymous==0) & (s_set.nucleotide!=s_set.consensus_nucleotide))/top)\n",
    "# Synonymous, reversion\n",
    "n_poly.append(np.sum((s_set.nonsynonymous==0) & (s_set.nucleotide==s_set.consensus_nucleotide))/top)\n",
    "# Nonsynonymous, reversion\n",
    "n_poly.append(np.sum((s_set.nonsynonymous> 0) & (s_set.in_epitope==False) & (s_set.glycan==0)\n",
    "                      & (s_set.exposed==False) & (s_set.flanking==0) \n",
    "                      & (s_set.nucleotide==s_set.consensus_nucleotide))/top)\n",
    "# Other nonsynonymous\n",
    "n_poly.append(1-np.sum(n_poly))\n",
    "\n",
    "\n",
    "# Get enrichment curves\n",
    "\n",
    "norm  = float(len(df))\n",
    "top1  = int(np.round(len(df)*0.01))\n",
    "top2  = int(np.round(len(df)*0.02))\n",
    "top20 = int(np.round(len(df)*0.20))\n",
    "\n",
    "x = []\n",
    "y_CD8_MPL = []\n",
    "y_CD8_SL  = []\n",
    "y_rev_MPL = []\n",
    "y_rev_SL  = []\n",
    "\n",
    "CD8_bg = 6803 / 350045\n",
    "rev_bg = 4265 / 350045\n",
    "CD8_rev_bg = (4383 - 4265) / 350045\n",
    "\n",
    "for n in range(top1, top20+1):\n",
    "    s_set_MPL = df.iloc[s_MPL_sorted[:n]]\n",
    "    s_set_SL  = df.iloc[s_SL_sorted[:n] ]\n",
    "    x.append(n/norm)\n",
    "    y_CD8_MPL.append(np.sum((s_set_MPL.nonsynonymous>0) & (s_set_MPL.in_epitope==True))/n)\n",
    "    y_CD8_SL.append( np.sum((s_set_SL.nonsynonymous >0) & (s_set_SL.in_epitope ==True))/n)\n",
    "    y_rev_MPL.append(np.sum((s_set_MPL.nonsynonymous>0) & (s_set_MPL.in_epitope==False) \n",
    "                            & (s_set_MPL.nucleotide==s_set_MPL.consensus_nucleotide))/n)\n",
    "    y_rev_SL.append( np.sum((s_set_SL.nonsynonymous >0) & (s_set_SL.in_epitope==False)\n",
    "                            & (s_set_SL.nucleotide ==s_set_SL.consensus_nucleotide ))/n)\n",
    "    \n",
    "y_CD8_MPL = np.array(y_CD8_MPL)/CD8_bg\n",
    "y_CD8_SL  = np.array(y_CD8_SL )/CD8_bg\n",
    "y_rev_MPL = np.array(y_rev_MPL)/rev_bg\n",
    "y_rev_SL  = np.array(y_rev_SL )/rev_bg\n",
    "\n",
    "s_set_MPL = df.iloc[s_MPL_sorted[:top1]]\n",
    "print('MPL enrichment in escape mutations that are reversions (top 1%%): %d' % \n",
    "      (np.round(np.sum((s_set_MPL.nonsynonymous>0) & (s_set_MPL.in_epitope==True) \n",
    "             & (s_set_MPL.nucleotide==s_set_MPL.consensus_nucleotide))/(top1 * CD8_rev_bg))))\n",
    "print('(MPL, SL) enrichment in escape mutations (top 1%%): (%.0f, %.0f)'       % (y_CD8_MPL[0], y_CD8_SL[0]))\n",
    "print('\\tMPL fraction of SL: %.2f' % ((y_CD8_MPL[0])/y_CD8_SL[0]))\n",
    "print('(MPL, SL) enrichment in non-epitope reversions (top 1%%): (%.0f, %.0f)' % (y_rev_MPL[0], y_rev_SL[0]))\n",
    "print('\\tMPL fraction of SL: %.2f' % ((y_rev_MPL[0])/y_rev_SL[0]))\n",
    "print('MPL percentage gain in escape mutations (top 1%%): %.0f' % (100*(y_CD8_MPL[0] - y_CD8_SL[0])/y_CD8_SL[0]))\n",
    "\n",
    "\n",
    "# Pass information to figure generator\n",
    "\n",
    "pdata = {\n",
    "    'n_poly':    n_poly,     # fraction of top 1% most beneficial mutations in different categories\n",
    "    'x_enr':     x,          # percentile values for enrichment analysis\n",
    "    'y_CD8_MPL': y_CD8_MPL,  # enrichment in putative CD8+ T cell escape mutations, MPL\n",
    "    'y_CD8_SL':  y_CD8_SL,   # enrichment in putative CD8+ T cell escape mutations, SL\n",
    "    'y_rev_MPL': y_rev_MPL,  # enrichment in nonsynonymous reversions, MPL\n",
    "    'y_rev_SL':  y_rev_SL,   # enrichment in nonsynonymous reversions, SL\n",
    "}\n",
    "\n",
    "fig.plot_figure_hiv_summary(**pdata)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='figs3-figs6'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Supplementary Figure 3. Distributions of total effects on inferred selection coefficients due to linkage.  \n",
    "### Supplementary Figure 6. Distributions of linkage effects on inferred selection, based on genomic distance."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "706010164-5\t23T\t0.590\n",
      "706010164-5\t44T\t0.614\n",
      "705010198-3\t194G\t0.510\n",
      "705010185-3\t225T\t0.760\n",
      "705010162-3\t101G\t0.761\n",
      "705010162-3\t266G\t0.466\n",
      "704010042-5\t29A\t0.504\n",
      "704010042-5\t30A\t0.549\n",
      "704010042-3\t96G\t0.864\n",
      "704010042-3\t874-\t0.803\n",
      "703010256-5\t9A\t0.874\n",
      "703010256-5\t172T\t0.621\n",
      "703010256-3\t62A\t0.601\n",
      "703010256-3\t353A\t0.684\n",
      "703010159-5\t3G\t0.406\n",
      "703010159-5\t34G\t0.488\n",
      "703010159-5\t179A\t0.436\n",
      "703010159-3\t370T\t0.637\n",
      "703010159-3\t371A\t0.725\n",
      "703010159-3\t399G\t0.509\n",
      "703010159-3\t403G\t0.658\n",
      "703010159-3\t424G\t0.488\n",
      "703010131-5\t3G\t0.582\n",
      "703010131-3\t31C\t0.461\n",
      "703010131-3\t32T\t0.402\n",
      "703010131-3\t52T\t0.465\n",
      "703010131-3\t126T\t0.446\n",
      "703010131-3\t263A\t0.832\n",
      "703010131-3\t352A\t1.072\n",
      "703010131-3\t602A\t0.684\n",
      "703010131-3\t603G\t0.548\n",
      "703010131-3\t605T\t0.888\n",
      "703010131-3\t619G\t0.827\n",
      "703010131-3\t620C\t1.281\n",
      "703010131-3\t620T\t0.422\n",
      "703010131-3\t621G\t0.440\n",
      "703010131-3\t624G\t0.532\n",
      "700010607-3\t199A\t0.443\n",
      "700010470-3\t1T\t0.612\n",
      "700010470-3\t10T\t0.551\n",
      "700010470-3\t352T\t0.517\n",
      "700010470-3\t355G\t0.905\n",
      "700010077-3\t35C\t0.802\n",
      "700010077-3\t87A\t0.494\n",
      "700010077-3\t186C\t0.766\n",
      "700010058-5\t16A\t0.526\n",
      "700010040-5\t14G\t0.410\n",
      "700010040-5\t31C\t0.557\n",
      "700010040-5\t95C\t0.449\n",
      "700010040-3\t40G\t0.578\n",
      "700010040-3\t41A\t0.553\n",
      "700010040-3\t103A\t0.563\n",
      "700010040-3\t270A\t0.405\n",
      "cap256-3\t99A\t0.494\n",
      "cap256-3\t159C\t0.419\n",
      "cap256-3\t159G\t0.470\n",
      "cap256-3\t162T\t0.491\n",
      "cap256-3\t185C\t0.502\n",
      "cap256-3\t294A\t0.614\n",
      "cap256-3\t312G\t0.833\n",
      "cap256-3\t324A\t0.453\n",
      "cap256-3\t536A\t0.423\n",
      "cap256-3\t657A\t0.407\n",
      "cap256-3\t742A\t0.465\n",
      "Delta s histogram done.\n",
      "Delta s distance distribution done.\n"
     ]
    }
   ],
   "source": [
    "from imp import reload\n",
    "reload(mp)\n",
    "reload(fig)\n",
    "\n",
    "# Iterate through patients/regions and get |Delta s| values\n",
    "\n",
    "df_range = pd.read_csv('%s/range.csv' % (HIV_DIR), comment='#', memory_map=True)\n",
    "ds_patient = []\n",
    "ds_region = []\n",
    "sum_ds_values = []\n",
    "full_ds_values = []\n",
    "full_ds_distance = []\n",
    "large_sum_ds = []\n",
    "\n",
    "for it, entry in df_range.iterrows():\n",
    "    tag = str(entry.tag)\n",
    "    \n",
    "    df_ds = pd.read_csv('%s/analysis/%s-delta-s.csv' % (HIV_DIR, tag), comment='#', memory_map=True)\n",
    "    df_ds = df_ds[~((df_ds.mask_polymorphic_index==df_ds.target_polymorphic_index) \n",
    "                     & (df_ds.mask_nucleotide==df_ds.target_nucleotide))]\n",
    "    site_ids = np.unique(df_ds.mask_polymorphic_index)\n",
    "    temp_sum_ds_values = []\n",
    "    for i in site_ids:\n",
    "        df_site = df_ds[df_ds.mask_polymorphic_index==i]\n",
    "        var_ids = np.unique(df_site.mask_nucleotide)\n",
    "        for a in var_ids:\n",
    "            df_var = df_site[df_site.mask_nucleotide==a]\n",
    "            sum_ds = np.sum(np.fabs(df_var.effect))\n",
    "            temp_sum_ds_values.append(sum_ds)\n",
    "            full_ds_values += list(np.fabs(df_var.effect))\n",
    "            full_ds_distance += list(df_var.distance)\n",
    "            \n",
    "            if sum_ds>0.4:\n",
    "                large_sum_ds.append([tag, i, a])\n",
    "                print('%s\\t%d%s\\t%.3f' % (tag, i, a, sum_ds))\n",
    "    \n",
    "    temp_sum_ds_values = np.array(temp_sum_ds_values)\n",
    "    \n",
    "    ds_patient.append(entry.ppt)\n",
    "    ds_region.append(str(entry.tag)[-1])\n",
    "    sum_ds_values.append(temp_sum_ds_values)\n",
    "    \n",
    "# Pass information to figure generator\n",
    "\n",
    "pdata = {\n",
    "    'patient_list': ds_patient,\n",
    "    'region_list':  ds_region,\n",
    "    'ds_values':    sum_ds_values,\n",
    "    'fig_title':    'figs3-absolute-delta-s'\n",
    "}\n",
    "\n",
    "fig.plot_supplementary_figure_absolute_delta_s(**pdata)\n",
    "\n",
    "\n",
    "# Pass information to figure generator\n",
    "\n",
    "pdata = {\n",
    "    'ds_values':   full_ds_values,\n",
    "    'ds_distance': full_ds_distance,\n",
    "    'fig_title':   'figs6-delta-s-distance'\n",
    "}\n",
    "\n",
    "fig.plot_supplementary_figure_delta_s_distance(**pdata)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='figs4'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Supplementary Figure 4. Visualization of particularly large contributions to inferred selection due to linkage.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['706010164', '706010164', '705010198', '705010198', '705010185', '705010185', '705010162', '705010162', '704010042', '704010042', '703010256', '703010256', '703010159', '703010159', '703010131', '703010131', '700010607', '700010607', '700010470', '700010470', '700010077', '700010077', '700010058', '700010058', '700010040', '700010040', 'cap256']\n",
      "['5', '3', '5', '3', '5', '3', '5', '3', '5', '3', '5', '3', '5', '3', '5', '3', '5', '3', '5', '3', '5', '3', '5', '3', '5', '3', '3']\n",
      "706010164-5 ['GQVDCSPGIW', 'EAMSQANNA', 'PFRDYVDRF', 'KEGHIARNCKA']\n",
      "706010164-3 ['EEVGFPVRPQV', 'KRQDILDLWVY', 'NCYCKMCSY', 'SFDPIPIHY', 'EEIIIRSENL']\n",
      "705010198-5 ['TSTLQEQVAW']\n",
      "705010198-3 ['KAAFDLSFF']\n",
      "705010185-5 ['TPQDLNTML', 'GTEELRSLY']\n",
      "705010185-3 []\n",
      "705010162-5 ['GKEGHIAKN', 'VSRGIRKVL', 'VHKGIKVKD', 'EEMNLTGKW']\n",
      "705010162-3 ['EILDLWVYH', 'LQAVRIIKI', 'HYFDCFAGS', 'KPQVPLRPM', 'RIRKTAPTA']\n",
      "704010042-5 ['QMVHQALSP', 'NETPGIRYQ']\n",
      "704010042-3 ['DETLLQAVR', 'KQRVHALFY', 'NYTDIIYRL', 'REILDLWVY']\n",
      "703010256-5 ['NRETKMGKA', 'PIQLPEKDS', 'QMVHQPLSPR', 'DRFFKTLRA']\n",
      "703010256-3 ['LVQDWGLEL', 'EDRWNKPQK', 'QLAHRHMAR', 'EPIDPNLEPW', 'RNRSIRLVN', 'TAVPWDSSW']\n",
      "703010159-5 ['TPQDLNTML', 'NWMTDTLLI']\n",
      "703010159-3 ['FPRPWLHNL', 'PTEPVPFQL']\n",
      "703010131-5 ['VKVIEEKAF', 'RKAKIIKDY']\n",
      "703010131-3 ['SPLSFQTLI', 'EEVGFPVKPQV', 'RKAKIIKDY', 'VTVYYGVPV', 'KTACNNCYC', 'FQKKGLGISY', 'CPKISFDPI']\n",
      "700010607-3 []\n",
      "700010470-5 ['ELYPMTSLK', 'RGRQKVVSL', 'QIYPGIKVK', 'GGKKKYQLK', 'DIKDTKEAL']\n",
      "700010470-3 ['RIKTWKSLVK', 'VPLRPMTFK', 'RQGLERALI', 'RLAYQHIAKEK', 'LVAPKKTKP']\n",
      "700010077-5 ['DEPAAVGVG', 'ASRELERF', 'FYKTLRAEQ', 'RMYSPTSIL', 'TSTLQEQVGW']\n",
      "700010077-3 ['AVLNIPTRI', 'QF-RNKTIVF', 'DEPAAVGVG', 'KAALDLSHF', 'DLLKTVRLI', 'TTVPWNVSW', 'TLSHVVDKL', 'DRVIEELQR', 'VAREIHPEF']\n",
      "700010058-5 ['ISPRTLNAW', 'TSTLQEQIGW']\n",
      "700010058-3 ['HTQGYFPDW', 'ERYLRDQQL']\n",
      "700010040-5 ['KELYPLASL', 'CGKEGHIAR', 'EEMNLPGRW', 'HTDNGSNFT']\n",
      "700010040-3 ['RACRAILHI', 'AIRKAILGR', 'RLRDLLLIV', 'SSLAFRHVAR']\n",
      "cap256-3 []\n",
      "Hive plots done.\n"
     ]
    }
   ],
   "source": [
    "from imp import reload\n",
    "reload(mp)\n",
    "reload(fig)\n",
    "\n",
    "print(ds_patient)\n",
    "print(ds_region)\n",
    "\n",
    "ds_patient = ['706010164', '706010164', '705010198', \n",
    "              '705010198', '705010185', '705010185', \n",
    "              '705010162', '705010162', '704010042', '704010042', '703010256', \n",
    "              '703010256', '703010159', '703010159', '703010131', '703010131', \n",
    "              '700010607', '700010470', '700010470', '700010077', \n",
    "              '700010077', '700010058', '700010058', '700010040', '700010040', 'cap256']\n",
    "\n",
    "ds_region = ['5', '3', '5',\n",
    "             '3', '5', '3', \n",
    "             '5', '3', '5', '3', '5',\n",
    "             '3', '5', '3', '5', '3',\n",
    "             '3', '5', '3', '5',\n",
    "             '3', '5', '3', '5', '3', '3']\n",
    "\n",
    "pdata = {\n",
    "    'patient_list': ds_patient,\n",
    "    'region_list':  ds_region,\n",
    "    'fig_title':    'figs4-delta-s-hive',\n",
    "}\n",
    "\n",
    "fig.plot_supplementary_figure_delta_s_hive(**pdata)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='figs5'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Supplementary Figure 5. Distribution of the maximum observed change in frequency per day for highly influential variants versus other variants."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "706010164-5\t23T\t0.008\t\n",
      "706010164-5\t44T\t0.010\tEPITOPE\n",
      "705010198-3\t194G\t0.055\tEPITOPE\n",
      "705010185-3\t225T\t0.015\t\n",
      "705010162-3\t101G\t0.039\t\n",
      "705010162-3\t266G\t0.022\t\n",
      "704010042-5\t29A\t0.013\tEPITOPE\n",
      "704010042-5\t30A\t0.026\tEPITOPE\n",
      "704010042-3\t96G\t0.037\tEPITOPE\n",
      "704010042-3\t874-\t0.037\t\n",
      "703010256-5\t9A\t0.006\t\n",
      "703010256-5\t172T\t0.004\tEPITOPE\n",
      "703010256-3\t62A\t0.018\t\n",
      "703010256-3\t353A\t0.006\t\n",
      "703010159-5\t3G\t0.005\t\n",
      "703010159-5\t34G\t0.028\t\n",
      "703010159-5\t179A\t0.028\t\n",
      "703010159-3\t370T\t0.020\tEPITOPE\n",
      "703010159-3\t371A\t0.011\tEPITOPE\n",
      "703010159-3\t399G\t0.050\t\n",
      "703010159-3\t403G\t0.029\t\n",
      "703010159-3\t424G\t0.020\t\n",
      "703010131-5\t3G\t0.008\t\n",
      "703010131-3\t31C\t0.006\t\n",
      "703010131-3\t32T\t0.097\t\n",
      "703010131-3\t52T\t0.009\t\n",
      "703010131-3\t126T\t0.012\t\n",
      "703010131-3\t263A\t0.014\t\n",
      "703010131-3\t352A\t0.011\t\n",
      "703010131-3\t602A\t0.069\t\n",
      "703010131-3\t603G\t0.069\t\n",
      "703010131-3\t605T\t0.014\t\n",
      "703010131-3\t619G\t0.049\tEPITOPE\n",
      "703010131-3\t620C\t0.028\tEPITOPE\n",
      "703010131-3\t620T\t0.003\tEPITOPE\n",
      "703010131-3\t621G\t0.021\tEPITOPE\n",
      "703010131-3\t624G\t0.021\tEPITOPE\n",
      "700010607-3\t199A\t0.026\t\n",
      "700010470-3\t1T\t0.034\tEPITOPE\n",
      "700010470-3\t10T\t0.018\t\n",
      "700010470-3\t352T\t0.010\tEPITOPE\n",
      "700010470-3\t355G\t0.020\tEPITOPE\n",
      "700010077-3\t35C\t0.048\tEPITOPE\n",
      "700010077-3\t87A\t0.052\tEPITOPE\n",
      "700010077-3\t186C\t0.037\tEPITOPE\n",
      "700010058-5\t16A\t0.007\tEPITOPE\n",
      "700010040-5\t14G\t0.004\t\n",
      "700010040-5\t31C\t0.008\tEPITOPE\n",
      "700010040-5\t95C\t0.012\t\n",
      "700010040-3\t40G\t0.008\t\n",
      "700010040-3\t41A\t0.030\t\n",
      "700010040-3\t103A\t0.012\t\n",
      "700010040-3\t270A\t0.007\t\n",
      "cap256-3\t99A\t0.017\t\n",
      "cap256-3\t159C\t0.002\t\n",
      "cap256-3\t159G\t0.003\t\n",
      "cap256-3\t162T\t0.006\t\n",
      "cap256-3\t185C\t0.023\t\n",
      "cap256-3\t294A\t0.006\t\n",
      "cap256-3\t312G\t0.019\t\n",
      "cap256-3\t324A\t0.003\t\n",
      "cap256-3\t536A\t0.010\t\n",
      "cap256-3\t657A\t0.011\t\n",
      "cap256-3\t742A\t0.011\t\n",
      "Highly influential variants in epitopes: 39.6%\n",
      "Delta x distribution done.\n"
     ]
    }
   ],
   "source": [
    "from imp import reload\n",
    "reload(mp)\n",
    "reload(fig)\n",
    "\n",
    "df_range     = pd.read_csv('%s/range.csv' % (HIV_DIR), comment='#', memory_map=True)\n",
    "dx_max_sweep = []\n",
    "dx_max_other = []\n",
    "in_epitope_sweep = []\n",
    "in_epitope_other = []\n",
    "xx = []\n",
    "xy = []\n",
    "\n",
    "for it, entry in df_range.iterrows():\n",
    "    tag = str(entry.tag)\n",
    "    \n",
    "    df_tr = pd.read_csv('%s/analysis/%s-poly.csv'    % (HIV_DIR, tag), comment='#', memory_map=True)\n",
    "    times = np.sort(np.array([c.split('_')[-1] for c in df_tr.columns if 'f_at' in c], int))\n",
    "    \n",
    "    df_ds = pd.read_csv('%s/analysis/%s-delta-s.csv' % (HIV_DIR, tag), comment='#', memory_map=True)\n",
    "    df_ds = df_ds[~((df_ds.mask_polymorphic_index==df_ds.target_polymorphic_index) \n",
    "                     & (df_ds.mask_nucleotide==df_ds.target_nucleotide))]\n",
    "    site_ids = np.unique(df_ds.mask_polymorphic_index)\n",
    "\n",
    "    for i in site_ids:\n",
    "        df_site = df_ds[df_ds.mask_polymorphic_index==i]\n",
    "        var_ids = np.unique(df_site.mask_nucleotide)\n",
    "        for a in var_ids:\n",
    "            df_var = df_site[df_site.mask_nucleotide==a]\n",
    "            sum_ds = np.sum(np.fabs(df_var.effect))\n",
    "            dx_max = 0\n",
    "            \n",
    "            df_tr_var = df_tr[(df_tr.polymorphic_index==i) & (df_tr.nucleotide==a)].iloc[0]\n",
    "            for k in range(len(times)-1):\n",
    "                dx = np.fabs(df_tr_var['f_at_%d'%times[k+1]] - df_tr_var['f_at_%d'%times[k]]) / (times[k+1]-times[k])\n",
    "                if dx>dx_max:\n",
    "                    dx_max = dx\n",
    "            \n",
    "            if sum_ds>0.4:\n",
    "                dx_max_sweep.append(dx_max)\n",
    "                if 'cap256' not in tag:\n",
    "                    if pd.notnull(df_tr_var.epitope):\n",
    "                        in_epitope_sweep.append(1)\n",
    "                    else:\n",
    "                        in_epitope_sweep.append(0)\n",
    "                    \n",
    "                print('%s\\t%d%s\\t%.3f\\t%s' % (tag, i, a, dx_max, 'EPITOPE' if pd.notnull(df_tr_var.epitope) else ''))\n",
    "                \n",
    "            else:\n",
    "                dx_max_other.append(dx_max)\n",
    "                if 'cap256' not in tag:\n",
    "                    if pd.notnull(df_tr_var.epitope):\n",
    "                        in_epitope_other.append(1)\n",
    "                    else:\n",
    "                        in_epitope_other.append(0)\n",
    "\n",
    "print('Highly influential variants in epitopes: %.1f%%' % (100 * np.sum(in_epitope_sweep)/len(in_epitope_sweep)))\n",
    "\n",
    "pdata = {\n",
    "    'dx_sweep':  dx_max_sweep,\n",
    "    'dx_other':  dx_max_other,\n",
    "    'fig_title': 'figs5-max-dx',\n",
    "}\n",
    "\n",
    "fig.plot_supplementary_figure_max_dx(**pdata)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='fig3'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Figure 3. Escape from the KF9 epitope targeted by individual CH77."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.0000\t-0.0005\t0.0048\t0.0007\t0.0020\t0.0158\t-0.0107\t0.0022\t-0.0053\n",
      "-0.0023\t0.0000\t0.0036\t0.0002\t0.0008\t0.0077\t-0.0062\t0.0011\t0.0026\n",
      "0.0209\t0.0035\t0.0000\t0.0002\t0.0007\t0.0011\t-0.0027\t-0.0066\t0.0000\n",
      "0.0064\t0.0005\t0.0005\t0.0000\t0.0027\t-0.0006\t-0.0018\t0.0003\t0.0010\n",
      "0.0098\t0.0009\t0.0008\t0.0014\t0.0000\t-0.0021\t-0.0032\t0.0006\t0.0008\n",
      "0.02088378\n",
      "700010077-3\n",
      "variant\ts\t(sigma)\n",
      "\n",
      "180/424 (42%) significant at 0.00 sigma\n",
      "\n",
      "GACCTGCTCAAGACAGTAAGATTAATC\n",
      "ACTTTAAGCCATGTAGTTGACAAATTA\n",
      "CAATTTAGG---AATAAAACAATAGTCTTT\n",
      "ACTACTGTGCCTTGGAATGTTAGTTGG\n",
      "GATAGGGTTATAGAAGAATTACAAAGG\n",
      "GCTGTTCTTAACATACCTACAAGAATA\n",
      "GATGAGCCAGCAGCAGTGGGGGTGGGA\n",
      "AAGGCAGCTCTTGATCTTAGCCACTTT\n",
      "GTAGCCCGAGAAATACATCCGGAGTTT\n",
      "Figure 3 done.\n"
     ]
    }
   ],
   "source": [
    "from imp import reload\n",
    "reload(mp)\n",
    "reload(fig)\n",
    "\n",
    "\n",
    "# Pass information to figure generator\n",
    "\n",
    "pdata = {\n",
    "    'patient':       '700010077',                    # patient\n",
    "    'region':        '3',                            # sequencing region in which the epitope is located\n",
    "    'inf_idxs':      [[35,'C'],[87,'A'],[124,'G'],   # locations of highly influential variants (polymorphic index)\n",
    "                      [149,'A']],                    # note: 186C = 9040C, so it is already included by default\n",
    "    'epitope':       'KAALDLSHF',                    # epitope sequence\n",
    "    'epitope_range': [[        6000,         6027],  #   rev: DLLKTVRLI  # CH077 epitopes\n",
    "                      [6225+(340*3), 6225+(349*3)],  # gp120: TLSHVVDKL\n",
    "                      [6225+(351*3), 6225+(361*3)],  # gp120: QF-RNKTIVF\n",
    "                      [7758+ (94*3), 7758+(103*3)],  #  gp41: TTVPWNVSW\n",
    "                      [7758+(315*3), 7758+(324*3)],  #  gp41: DRVIEELQR\n",
    "                      [7758+(327*3), 7758+(336*3)],  #  gp41: AVLNIPTRI\n",
    "                      [8797+ (25*3), 8797+ (31*3)],  #   nef: DEPAAVGVG (start is earlier but inserted wrt HXB2)\n",
    "                      [8797+ (81*3), 8797+ (90*3)],  #   nef: KAALDLSHF\n",
    "                      [8797+(193*3), 8797+(202*3)]], #   nef: VAREIHPEF\n",
    "    'epitope_label': ['DI9', 'TL9', 'QF9', 'TW9',    # epitope labels\n",
    "                      'DR9', 'AI9', 'DG9', 'KF9',\n",
    "                      'VF9'],\n",
    "    'cov_label':     'KF9',                          # plot covariance links for sites in this epitope\n",
    "    'label2ddr':     {'tat exon 2': 0.09,            # label adjustment for plotting\n",
    "                      'rev exon 2': 0.13,\n",
    "                      'QF9':        0.005,\n",
    "                      'AI9':        0.09,\n",
    "                      'TL9':        0.09,\n",
    "                      'DI9':        0.13,\n",
    "                      'DR9':        0.15  },\n",
    "    'legend_loc':    'bottom',                       # location of the circle legend\n",
    "    'traj_ticks':    [0, 80, 160],                   # tick marks for trajectory plot\n",
    "    'sel_ticks':     [0, 0.05, 0.10],                # tick marks for selection plot\n",
    "    'sel_minors':    [0.025, 0.075],                 # minor tick marks for selection plot\n",
    "    'sel_space':     0.02,                           # spacing between MPL and independent selection plots\n",
    "    'fig_title':     'fig3-ch77-kf9'                 # title of figure\n",
    "}\n",
    "\n",
    "fig.plot_figure_ch77_kf9(**pdata)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='figs7-figs8'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Supplementary Figure 7. A simple example of clonal interference (CH58 epitope TW10).  \n",
    "### Supplementary Figure 8. A complex example of clonal interference (CH131 epitope EV11)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "700010058-5\n",
      "variant\ts\t(sigma)\n",
      "\n",
      "77/192 (40%) significant at 0.00 sigma\n",
      "\n",
      "ATATCGCCTAGAACTTTAAATGCATGG\n",
      "ACTAGTACCCTTCAGGAACAAATAGGATGG\n",
      "Figures S7-S8 done.\n",
      "703010131-3\n",
      "variant\ts\t(sigma)\n",
      "\n",
      "379/1551 (24%) significant at 0.00 sigma\n",
      "\n",
      "AAGACTGCTTGTAATAATTGTTATTGT\n",
      "TTTCAGAAAAAAGGCTTAGGCATTTCCTAT\n",
      "GTCACAGTCTATTATGGGGTACCTGTA\n",
      "TGTCCAAAGATCTCGTTTGATCCGATT\n",
      "TCACCTTTATCGTTCCAGACCCTTATC\n",
      "GAGGTAGGCTTTCCAGTCAAACCCCAGGTG\n",
      "Figures S7-S8 done.\n"
     ]
    }
   ],
   "source": [
    "from imp import reload\n",
    "reload(mp)\n",
    "reload(fig)\n",
    "\n",
    "# Pass information to figure generator\n",
    "\n",
    "pdata = {\n",
    "    'patient':       '700010058',\n",
    "    'region':        '5',\n",
    "    'epitope':       'TSTLQEQIGW',\n",
    "    'epitope_range': [[1186+ (14*3), 1186+ (23*3)],  #   p24: ISPRTLNAW  # CH058 epitopes\n",
    "                      [1186+(107*3), 1186+(117*3)]], #   p24: TSTLQEQIGW\n",
    "    'epitope_label': ['IW9', 'TW10'],\n",
    "    'cov_label':     'TW10',\n",
    "    'label2ddr':     {},\n",
    "    'legend_loc':    'top', \n",
    "    'traj_ticks':    [0, 70, 140, 210, 280, 350],\n",
    "    'sel_ticks':     [0, 0.06, 0.12],\n",
    "    'sel_minors':    [0.03, 0.09],\n",
    "    'sel_space':     0.06,  \n",
    "    'fig_title':     'figs7-ch58-tw10'\n",
    "}\n",
    "fig.plot_supplementary_figure_epitope(**pdata)\n",
    "\n",
    "\n",
    "pdata = {\n",
    "    'patient':       '703010131',\n",
    "    'region':        '3',\n",
    "    'epitope':       'EEVGFPVKPQV',\n",
    "    'epitope_range': [[5885, 5911+1],  #   tat: KTACNNCYC  # CH131 epitopes\n",
    "                      [5942, 5971+1],  #   tat: FQKKGLGISY\n",
    "                      [6330, 6356+1],  # gp120: VTVYYGVPV\n",
    "                      [6837, 6863+1],  # gp120: CPKISFDPI\n",
    "                      [8361, 8387+1],  #  gp41: SPLSFQTLI\n",
    "                      [8989, 9018+1]], #   nef: EEVGFPVKPQV\n",
    "                      #[6062, 6073+1],  #   vpu: MIELLARVD    (LLARV inserted wrt HXB2)\n",
    "                      #[7329, 7355+1],  # gp120: LEITTHSFN\n",
    "                      #[9037, 9063+1],  #   nef: YKGAFDLSF\n",
    "                      #[9265, 9291+1]], #   nef: NSGENSCLL\n",
    "    'epitope_label': ['KC9', 'FY10', 'VV9', 'CI9', 'SI9', \n",
    "                      'EV11'],#, 'MD9', 'LN9', 'YF9', 'NL9'],\n",
    "    'cov_label':     'EV11',\n",
    "    'label2ddr':     {'KC9':        0.065,\n",
    "                      'rev exon 1': 0.225,\n",
    "                      'FY10':       0.11,\n",
    "                      'CI9':        0.005,\n",
    "                      'tat exon 2': 0.10,            \n",
    "                      'rev exon 2': 0.135,\n",
    "                      'SI9':        0.15,\n",
    "                      'EV11':       0.12  },\n",
    "    'legend_loc':    'top',\n",
    "    'traj_ticks':    [0, 70, 140, 210, 280, 350],\n",
    "    'sel_ticks':     [0, 0.05, 0.10],\n",
    "    'sel_minors':    [0.025, 0.075],\n",
    "    'sel_space':     0.02, \n",
    "    'fig_title':     'figs8-ch131-ev11'\n",
    "}\n",
    "fig.plot_supplementary_figure_epitope(**pdata)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='fig4-figs9'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Figure 4. HIV evolution in individual CAP256.  \n",
    "### Supplementary Figure 9. Recombination between primary and superinfecting strains in individual CAP256."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Percentile of the most strongly selected VRC26 resistance variant: 8.1%\n",
      "\n",
      "synonymous:\t0 (MPL)\t6 (SL)\t of 11 total\n",
      "variant\ts_MPL\n",
      "6709C\t0.041\n",
      "6717T\t-0.012\n",
      "6730C\t0.003\n",
      "6730G\t0.006\n",
      "6730T\t0.010\n",
      "6731T\t0.003\n",
      "\n",
      "cap256-3\n",
      "variant\ts\t(sigma)\n",
      "\n",
      "708/1923 (36%) significant at 0.00 sigma\n",
      "\n",
      "AATACAATCACAGAGGTAAGAGATAAGCAAAAGAA\n",
      "Figure 4 done.\n",
      "Figure S9 done.\n"
     ]
    }
   ],
   "source": [
    "from imp import reload\n",
    "reload(fig)\n",
    "reload(mp)\n",
    "\n",
    "# Pass information to figure generator\n",
    "\n",
    "pdata = {\n",
    "    'patient': 'cap256',     # patient\n",
    "    'region':  '3',          # sequencing region in which the epitope is located\n",
    "}\n",
    "\n",
    "df = pd.read_csv('%s/analysis/total-selection.csv' % (HIV_DIR), comment='#', memory_map=True)\n",
    "df = df[df.ppt!='cap256']\n",
    "print('Percentile of the most strongly selected VRC26 resistance variant: %.1f%%\\n' \n",
    "      % (100*np.sum(df.s_MPL>0.012)/float(len(df))))\n",
    "\n",
    "fig.plot_figure_cap256_vrc26(**pdata)\n",
    "fig.plot_supplementary_figure_cap256_recombination(**pdata)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='figs2'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Supplementary Figure 2. Performance comparison of MPL versus other selection inference methods."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "medium_complex\tNRMSE between nan and nan\n",
      "medium_simple\tNRMSE between nan and nan\n",
      "medium_complex\ttime between -2.84 and 3.99\n",
      "medium_simple\ttime between -2.69 and 5.12\n",
      "Figure S2 done.\n"
     ]
    }
   ],
   "source": [
    "from imp import reload\n",
    "reload(fig)\n",
    "reload(mp)\n",
    "\n",
    "# Define a function to retrieve performance from stored data\n",
    "\n",
    "def get_plot_data(test, df_dict, methods):\n",
    "    w      = 0.25\n",
    "    x_ben  = [i for i in range(len(methods))]\n",
    "    x_del  = [i for i in x_ben]\n",
    "    x_t    = [i for i in x_ben]\n",
    "    x_err  = [i for i in x_ben]\n",
    "    y_ben  = []\n",
    "    y_del  = []\n",
    "    y_t    = []\n",
    "    y_err  = []\n",
    "\n",
    "    for i in range(len(methods)):\n",
    "        method = methods[i]\n",
    "        if methods[i]=='EandR':\n",
    "            method = 'EandR'\n",
    "        elif methods[i]=='IM':\n",
    "            method = 'Det'\n",
    "        elif methods[i]=='WFABC':\n",
    "            method = 'ABC'\n",
    "        df_test = df_dict[methods[i]]\n",
    "        df_test = df_test[(df_test.ns==100) & (df_test.deltat==10) & (df_test.method==method)]\n",
    "        y_ben.append(np.mean(df_test.AUROC_ben))\n",
    "        y_del.append(np.mean(df_test.AUROC_del))\n",
    "        y_t.append(np.log10(np.mean(df_test.runtime)))\n",
    "        \n",
    "        dscols = [j for j in df_test.columns if 'ds' in j]\n",
    "        scols  = [j for j in df_test.columns if ('s' in j) and (j not in dscols+['ns'])]\n",
    "        y_num = np.sqrt(np.sum(np.array(df_test[dscols]**2),axis=1))\n",
    "        y_den = np.sqrt(np.sum([NB_VALS[test] * (SB_VALS[test]**2), ND_VALS[test] * (SD_VALS[test]**2)]))\n",
    "        if method=='LLS':\n",
    "            y_num = []\n",
    "            y_den = []\n",
    "            for j in range(len(df_test)):\n",
    "                _temp_num = 0\n",
    "                _temp_den = 0\n",
    "                for k in range(NB_VALS[test]):\n",
    "                    if not pd.isnull(df_test.iloc[j]['ds%d' % k]):\n",
    "                        _temp_num += df_test.iloc[j]['ds%d' % k]**2\n",
    "                        _temp_den += SB_VALS[test]**2\n",
    "                for k in range(NB_VALS[test],L_VALS[test]-ND_VALS[test]):\n",
    "                    if not pd.isnull(df_test.iloc[j]['ds%d' % k]):\n",
    "                        _temp_num += df_test.iloc[j]['ds%d' % k]**2\n",
    "                for k in range(L_VALS[test]-ND_VALS[test],L_VALS[test]):\n",
    "                    if not pd.isnull(df_test.iloc[j]['ds%d' % k]):\n",
    "                        _temp_num += df_test.iloc[j]['ds%d' % k]**2\n",
    "                        _temp_den += SD_VALS[test]**2\n",
    "                y_num.append(np.sqrt(_temp_num))\n",
    "                y_den.append(np.sqrt(_temp_den))\n",
    "        y_err.append(np.mean(np.array(y_num)/np.array(y_den)))\n",
    "        \n",
    "    return x_ben, x_del, y_ben, y_del, x_t, y_t, x_err, y_err\n",
    "\n",
    "\n",
    "# Choose the test data sets and get performance data\n",
    "\n",
    "pdata = dict(test_sets=['medium_complex', 'medium_simple'],# 'small_complex', 'small_simple'],\n",
    "             traj_file=['%s/wfsim_medium_complex_0_T310_ns100_dt10.dat' % (SIM_DIR),\n",
    "                        '%s/wfsim_medium_simple_0_T1000_ns100_dt10.dat' % (SIM_DIR)],\n",
    "             t_ticks=[[0,  75, 150, 225,  300],\n",
    "                      [0, 250, 500, 750, 1000]],\n",
    "                      #[0, 25, 50, 75],\n",
    "                      #[0, 50, 100, 150]],\n",
    "            n_ben=[], n_neu=[], n_del=[], x_ben=[], y_ben=[], x_del=[], y_del=[], \n",
    "            x_err=[], y_err=[], x_t=[], y_t=[])\n",
    "\n",
    "methods = ['MPL', 'FIT', 'LLS', 'CLEAR', 'EandR', 'ApproxWF', 'WFABC', 'IM']\n",
    "\n",
    "for t in pdata['test_sets']:\n",
    "    pdata['n_ben'].append(NB_VALS[t])\n",
    "    pdata['n_neu'].append(L_VALS[t]-NB_VALS[t]-ND_VALS[t])\n",
    "    pdata['n_del'].append(ND_VALS[t])\n",
    "    \n",
    "    df_dict   = {}\n",
    "    for m in methods:\n",
    "        if m=='SL':\n",
    "            df_dict[m] = pd.read_csv('%s/MPL_%s_collected_extended.csv.gz' % (SIM_DIR, t))\n",
    "        else:\n",
    "            df_dict[m] = pd.read_csv('%s/%s_%s_collected_extended.csv.gz' % (SIM_DIR, m, t))\n",
    "    \n",
    "    x_ben, x_del, y_ben, y_del, x_t, y_t, x_err, y_err = get_plot_data(t, df_dict, methods)\n",
    "    \n",
    "    pdata['x_ben'].append(x_ben)\n",
    "    pdata['y_ben'].append(y_ben)\n",
    "    pdata['x_del'].append(x_del)\n",
    "    pdata['y_del'].append(y_del)\n",
    "    pdata['x_err'].append(x_err)\n",
    "    pdata['y_err'].append(y_err)\n",
    "    pdata['x_t'].append(x_t)\n",
    "    pdata['y_t'].append(y_t)\n",
    "    \n",
    "# Pass information to figure generator\n",
    "    \n",
    "fig.plot_supplementary_figure_performance(**pdata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
