{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# HIV Analysis\n",
    "\n",
    "Here we combine HIV sequence data from the [Los Alamos National Laboratory HIV Sequence Database](www.hiv.lanl.gov) and immunological data to investigate HIV evolution across 14 individuals. This data is contained in the `data/HIV/` directory. Supplementary scripts for data processing are stored in `HIV.py`.\n",
    "\n",
    "We downloaded sequence data for all patients and aligned them with the [HXB2 reference sequence](https://www.hiv.lanl.gov/components/sequence/HIV/asearch/query_one.comp?se_id=K03455) and the corresponding [clade consensus sequence](https://www.hiv.lanl.gov/content/sequence/NEWALIGN/align.html#consensus) using the LANL tool [HIValign](https://www.hiv.lanl.gov/content/sequence/VIRALIGN/viralign.html). We recorded immunological data for these patients in `.csv` files for analysis.\n",
    "\n",
    "### Sequence processing\n",
    "\n",
    "In the scripts below, we first process the HIV sequence data by\n",
    "1. associating times with sequences and removing all sequences for which no collection time is available,\n",
    "2. clipping full-length sequences down to the heavily-sequenced half-genome regions,\n",
    "3. removing sequences with excess gaps in the targeted region, which primarily consist of sequences that cover the opposite half-genome region,\n",
    "4. removing sites that are >= 95% gaps,\n",
    "5. time-ordering sequences,\n",
    "6. removing time points that are > 300 days from the previous time point and/or have < 4 sequences available at that time point, in order to ensure minimum quality of the correlations,\n",
    "6. imputing ambiguous nucleotides with the most frequently observed nucleotide at the same site, and\n",
    "7. saving the data in a format readable by the MPL program.\n",
    "\n",
    "During this process we track the location of known T cell epitopes and exposed regions of Env, and we record the transmitted/founder and clade consensus nucleotides at each site. We also save a shell script `HIV/MPL/run.sh` for running MPL to infer the selection coefficients underlying each trajectory."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load libraries and define global variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This notebook was prepared using:\n",
      "python version 3.7.3 (default, Mar 27 2019, 09:23:15) \n",
      "[Clang 10.0.1 (clang-1001.0.46.3)]\n",
      "numpy version 1.16.4\n",
      "pandas version 0.24.2\n"
     ]
    }
   ],
   "source": [
    "print('This notebook was prepared using:')\n",
    "\n",
    "import os\n",
    "import sys\n",
    "print('python version %s' % sys.version)\n",
    "\n",
    "import numpy as np\n",
    "print('numpy version %s' % np.__version__)\n",
    "\n",
    "import pandas as pd\n",
    "print('pandas version %s' % pd.__version__)\n",
    "\n",
    "import HIV\n",
    "\n",
    "# GLOBAL VARIABLES\n",
    "\n",
    "NUC = ['-', 'A', 'C', 'G', 'T']\n",
    "PRO = ['-', 'A', 'R', 'N', 'D', 'C', 'Q', 'E', 'G', 'H',\n",
    "       'I', 'L', 'K', 'M', 'F', 'P', 'S', 'T', 'W', 'Y', 'V']\n",
    "REF = NUC[0]\n",
    "CONS_TAG = 'CONSENSUS'\n",
    "HXB2_TAG = 'B.FR.1983.HXB2-LAI-IIIB-BRU.K03455.19535'\n",
    "HIV_DIR  = 'data/HIV'\n",
    "MPL_DIR  = 'src/MPL'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Process sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "706010164-5\n",
      "\tselected 98 of 199 sequences with <200 gaps in excess of consensus\n",
      "\tremoved 2 of 4262 sites with >95% gaps\n",
      "\texchanged A for R in sequence 72, site 97\n",
      "\texchanged T for Y in sequence 37, site 224\n",
      "\texchanged C for Y in sequence 37, site 343\n",
      "\texchanged A for R in sequence 40, site 638\n",
      "\texchanged A for N in sequence 97, site 693\n",
      "\texchanged G for R in sequence 3, site 747\n",
      "\texchanged G for R in sequence 14, site 747\n",
      "\texchanged G for R in sequence 22, site 747\n",
      "\texchanged G for R in sequence 28, site 747\n",
      "\texchanged G for S in sequence 37, site 897\n",
      "\texchanged C for Y in sequence 66, site 1011\n",
      "\texchanged C for Y in sequence 65, site 1112\n",
      "\texchanged A for R in sequence 63, site 1119\n",
      "\texchanged C for Y in sequence 3, site 1190\n",
      "\texchanged A for R in sequence 4, site 1382\n",
      "\texchanged A for R in sequence 45, site 1491\n",
      "\texchanged T for Y in sequence 32, site 1646\n",
      "\texchanged A for R in sequence 44, site 2136\n",
      "\texchanged A for R in sequence 34, site 2302\n",
      "\texchanged T for N in sequence 88, site 2469\n",
      "\texchanged G for R in sequence 90, site 2542\n",
      "\texchanged T for K in sequence 90, site 2550\n",
      "\texchanged A for N in sequence 96, site 2671\n",
      "\texchanged T for Y in sequence 22, site 3145\n",
      "\texchanged G for R in sequence 23, site 3357\n",
      "\texchanged C for Y in sequence 13, site 3407\n",
      "\texchanged C for Y in sequence 40, site 3664\n",
      "\texchanged A for R in sequence 35, site 3749\n",
      "\texchanged A for N in sequence 90, site 4072\n",
      "\texchanged C for Y in sequence 81, site 4118\n",
      "\texchanged G for N in sequence 80, site 4254\n",
      "\tmutant at site 4259 in codon that does not terminate in alignment, assuming syn\n",
      "\tmutant at site 4259 in codon that does not terminate in alignment, assuming syn\n",
      "706010164-3\n",
      "\tselected 102 of 199 sequences with <200 gaps in excess of consensus\n",
      "\tremoved 37 of 4678 sites with >95% gaps\n",
      "\texchanged T for Y in sequence 35, site 114\n",
      "\texchanged A for R in sequence 33, site 251\n",
      "\texchanged A for N in sequence 16, site 282\n",
      "\texchanged T for N in sequence 16, site 342\n",
      "\texchanged A for R in sequence 97, site 1045\n",
      "\texchanged A for R in sequence 14, site 1186\n",
      "\texchanged G for R in sequence 92, site 1302\n",
      "\texchanged C for Y in sequence 91, site 1384\n",
      "\texchanged G for N in sequence 16, site 1611\n",
      "\texchanged A for R in sequence 90, site 1698\n",
      "\texchanged T for Y in sequence 35, site 1726\n",
      "\texchanged C for Y in sequence 22, site 1837\n",
      "\texchanged G for R in sequence 86, site 1881\n",
      "\texchanged C for Y in sequence 84, site 1943\n",
      "\texchanged A for R in sequence 34, site 2324\n",
      "\texchanged C for Y in sequence 89, site 2483\n",
      "\texchanged C for Y in sequence 92, site 2623\n",
      "\texchanged C for Y in sequence 9, site 2672\n",
      "\texchanged G for R in sequence 92, site 3614\n",
      "\texchanged A for R in sequence 93, site 4008\n",
      "\texchanged A for R in sequence 81, site 4016\n",
      "\texchanged C for Y in sequence 92, site 4035\n",
      "\texchanged A for R in sequence 101, site 4311\n",
      "\texchanged T for Y in sequence 84, site 4427\n",
      "\texchanged T for K in sequence 92, site 4588\n",
      "\tmutant at site 0 in codon that does not terminate in alignment, assuming syn\n",
      "\tmutant at site 0 in codon that does not terminate in alignment, assuming syn\n",
      "\tmutant at site 1 in codon that does not terminate in alignment, assuming syn\n",
      "\tmutant at site 1 in codon that does not terminate in alignment, assuming syn\n",
      "705010198-5\n",
      "\tselected 57 of 119 sequences with <200 gaps in excess of consensus\n",
      "\tremoved 2 of 4256 sites with >95% gaps\n",
      "\texchanged C for Y in sequence 58, site 1202\n",
      "\texchanged G for R in sequence 11, site 1512\n",
      "\texchanged G for R in sequence 58, site 2029\n",
      "\texchanged A for R in sequence 49, site 2347\n",
      "\texchanged G for R in sequence 58, site 2350\n",
      "\texchanged A for R in sequence 2, site 2969\n",
      "\texchanged A for R in sequence 12, site 3034\n",
      "\texchanged A for N in sequence 13, site 3789\n",
      "\texchanged A for R in sequence 52, site 3885\n",
      "\texchanged G for R in sequence 58, site 4021\n",
      "\texchanged A for N in sequence 13, site 4081\n",
      "\texchanged A for N in sequence 13, site 4083\n",
      "\texchanged A for N in sequence 13, site 4084\n",
      "\texchanged T for Y in sequence 21, site 4188\n",
      "\tmutant at site 4253 in codon that does not terminate in alignment, assuming syn\n",
      "\tmutant at site 4253 in codon that does not terminate in alignment, assuming syn\n",
      "705010198-3\n",
      "\tselected 63 of 119 sequences with <200 gaps in excess of consensus\n",
      "\tremoved 38 of 4693 sites with >95% gaps\n",
      "\texchanged G for R in sequence 19, site 608\n",
      "\texchanged A for M in sequence 25, site 2499\n",
      "705010185-5\n",
      "\tselected 68 of 164 sequences with <200 gaps in excess of consensus\n",
      "\tremoved 11 of 4310 sites with >95% gaps\n",
      "\texchanged C for N in sequence 63, site 167\n",
      "\texchanged T for N in sequence 63, site 168\n",
      "\texchanged T for N in sequence 63, site 170\n",
      "\texchanged C for N in sequence 56, site 251\n",
      "\texchanged T for N in sequence 56, site 253\n",
      "\texchanged T for N in sequence 56, site 255\n",
      "\texchanged A for N in sequence 56, site 256\n",
      "\texchanged T for N in sequence 56, site 260\n",
      "\texchanged T for N in sequence 56, site 262\n",
      "\texchanged A for N in sequence 56, site 265\n",
      "\texchanged C for N in sequence 56, site 282\n",
      "\texchanged C for N in sequence 66, site 282\n",
      "\texchanged A for N in sequence 56, site 288\n",
      "\texchanged A for N in sequence 56, site 304\n",
      "\texchanged A for N in sequence 56, site 323\n",
      "\texchanged C for N in sequence 56, site 326\n",
      "\texchanged A for R in sequence 28, site 572\n",
      "\texchanged C for M in sequence 27, site 613\n",
      "\texchanged T for Y in sequence 52, site 629\n",
      "\texchanged C for Y in sequence 13, site 1184\n",
      "\texchanged A for R in sequence 9, site 1358\n",
      "\texchanged G for R in sequence 27, site 1472\n",
      "\texchanged A for R in sequence 9, site 1726\n",
      "\texchanged A for R in sequence 15, site 1744\n",
      "\texchanged T for Y in sequence 20, site 1800\n",
      "\texchanged A for W in sequence 8, site 3819\n",
      "\texchanged G for R in sequence 27, site 3997\n",
      "\texchanged A for R in sequence 52, site 4007\n",
      "\texchanged G for N in sequence 59, site 4294\n",
      "705010185-3\n",
      "\tselected 97 of 164 sequences with <200 gaps in excess of consensus\n",
      "\tremoved 50 of 4723 sites with >95% gaps\n",
      "\texchanged A for R in sequence 31, site 40\n",
      "\texchanged A for R in sequence 12, site 186\n",
      "\texchanged A for R in sequence 12, site 659\n",
      "\texchanged C for Y in sequence 28, site 755\n",
      "\texchanged G for R in sequence 19, site 936\n",
      "\texchanged A for R in sequence 14, site 1833\n",
      "\texchanged A for R in sequence 12, site 2961\n",
      "\texchanged A for R in sequence 11, site 2969\n",
      "\texchanged C for Y in sequence 11, site 3063\n",
      "\texchanged A for R in sequence 6, site 3143\n",
      "\texchanged A for R in sequence 7, site 3260\n",
      "\texchanged A for R in sequence 26, site 3304\n",
      "\texchanged T for Y in sequence 35, site 3702\n",
      "\texchanged A for R in sequence 35, site 3799\n",
      "\texchanged T for Y in sequence 39, site 3963\n",
      "\texchanged C for Y in sequence 8, site 3988\n",
      "\texchanged A for R in sequence 14, site 4322\n",
      "705010162-5\n",
      "\tselected 60 of 149 sequences with <200 gaps in excess of consensus\n",
      "\tremoved 2 of 4277 sites with >95% gaps\n",
      "\texchanged A for W in sequence 3, site 1007\n",
      "\texchanged A for R in sequence 16, site 1026\n",
      "\texchanged A for R in sequence 12, site 1259\n",
      "\texchanged A for R in sequence 6, site 1820\n",
      "\texchanged A for N in sequence 61, site 2170\n",
      "\texchanged A for R in sequence 14, site 2374\n",
      "\texchanged T for N in sequence 60, site 3642\n",
      "\texchanged A for M in sequence 16, site 3775\n",
      "\texchanged G for R in sequence 53, site 4148\n",
      "\texchanged G for R in sequence 53, site 4157\n",
      "\texchanged G for K in sequence 53, site 4211\n",
      "\tmutant at site 4274 in codon that does not terminate in alignment, assuming syn\n",
      "\tmutant at site 4274 in codon that does not terminate in alignment, assuming syn\n",
      "705010162-3\n",
      "\tselected 69 of 149 sequences with <200 gaps in excess of consensus\n",
      "\tremoved 70 of 4735 sites with >95% gaps\n",
      "\texchanged A for R in sequence 4, site 153\n",
      "\texchanged G for K in sequence 30, site 252\n",
      "\texchanged G for K in sequence 13, site 265\n",
      "\texchanged C for Y in sequence 64, site 1036\n",
      "\texchanged T for K in sequence 5, site 1478\n",
      "\texchanged A for R in sequence 13, site 1917\n",
      "\texchanged A for R in sequence 52, site 1944\n",
      "\texchanged G for K in sequence 64, site 2714\n",
      "\texchanged C for Y in sequence 64, site 3366\n",
      "\texchanged G for R in sequence 30, site 4033\n",
      "\texchanged A for R in sequence 61, site 4479\n",
      "704010042-5\n",
      "\tselected 85 of 179 sequences with <200 gaps in excess of consensus\n",
      "\tremoved 26 of 4280 sites with >95% gaps\n",
      "\texchanged A for R in sequence 73, site 68\n",
      "\texchanged A for R in sequence 28, site 340\n",
      "\texchanged A for R in sequence 43, site 340\n",
      "\texchanged G for R in sequence 17, site 438\n",
      "\texchanged A for R in sequence 44, site 691\n",
      "\texchanged A for R in sequence 14, site 886\n",
      "\texchanged A for R in sequence 43, site 1337\n",
      "\texchanged A for R in sequence 25, site 1543\n",
      "\texchanged A for R in sequence 74, site 1569\n",
      "\texchanged A for N in sequence 63, site 2113\n",
      "\texchanged C for N in sequence 53, site 2270\n",
      "\texchanged A for R in sequence 12, site 2329\n",
      "\texchanged T for Y in sequence 22, site 2544\n",
      "\texchanged G for R in sequence 22, site 2561\n",
      "\texchanged T for N in sequence 61, site 2934\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\texchanged T for W in sequence 17, site 3565\n",
      "\texchanged T for Y in sequence 49, site 3830\n",
      "\texchanged G for R in sequence 82, site 4017\n",
      "\texchanged A for R in sequence 45, site 4045\n",
      "\tmutant at site 4253 in codon that does not terminate in alignment, assuming syn\n",
      "\tmutant at site 4253 in codon that does not terminate in alignment, assuming syn\n",
      "704010042-3\n",
      "\tselected 93 of 179 sequences with <200 gaps in excess of consensus\n",
      "\tremoved 62 of 4714 sites with >95% gaps\n",
      "\texchanged A for R in sequence 37, site 579\n",
      "\texchanged T for W in sequence 14, site 767\n",
      "\texchanged T for Y in sequence 23, site 899\n",
      "\texchanged A for R in sequence 48, site 1080\n",
      "\texchanged C for Y in sequence 14, site 1092\n",
      "\texchanged T for Y in sequence 44, site 1377\n",
      "\texchanged G for R in sequence 14, site 2530\n",
      "\texchanged A for R in sequence 69, site 2764\n",
      "\texchanged A for R in sequence 10, site 2896\n",
      "\texchanged T for Y in sequence 42, site 3463\n",
      "\texchanged A for R in sequence 16, site 3617\n",
      "\texchanged T for Y in sequence 44, site 3685\n",
      "\texchanged C for M in sequence 23, site 3692\n",
      "\texchanged G for R in sequence 60, site 4108\n",
      "\texchanged G for R in sequence 15, site 4142\n",
      "\texchanged A for R in sequence 37, site 4412\n",
      "703010256-5\n",
      "\tselected 110 of 209 sequences with <200 gaps in excess of consensus\n",
      "\tremoved 26 of 4301 sites with >95% gaps\n",
      "\texchanged A for R in sequence 16, site 155\n",
      "\texchanged A for R in sequence 96, site 344\n",
      "\texchanged T for Y in sequence 86, site 441\n",
      "\texchanged T for Y in sequence 66, site 595\n",
      "\texchanged A for R in sequence 20, site 608\n",
      "\texchanged A for R in sequence 20, site 779\n",
      "\texchanged A for R in sequence 86, site 935\n",
      "\texchanged A for N in sequence 68, site 958\n",
      "\texchanged A for M in sequence 57, site 1388\n",
      "\texchanged A for R in sequence 61, site 1531\n",
      "\texchanged C for Y in sequence 21, site 1881\n",
      "\texchanged G for R in sequence 21, site 1981\n",
      "\texchanged G for R in sequence 55, site 2158\n",
      "\texchanged A for R in sequence 102, site 2529\n",
      "\texchanged T for Y in sequence 64, site 2537\n",
      "\texchanged A for R in sequence 13, site 2918\n",
      "\texchanged T for Y in sequence 55, site 2961\n",
      "\texchanged A for R in sequence 10, site 3118\n",
      "\texchanged A for R in sequence 65, site 3256\n",
      "\texchanged A for R in sequence 55, site 3391\n",
      "\texchanged T for Y in sequence 13, site 3758\n",
      "\tmutant at site 4274 in codon that does not terminate in alignment, assuming syn\n",
      "\tmutant at site 4274 in codon that does not terminate in alignment, assuming syn\n",
      "703010256-3\n",
      "\tselected 99 of 209 sequences with <200 gaps in excess of consensus\n",
      "\tremoved 59 of 4693 sites with >95% gaps\n",
      "\texchanged A for R in sequence 78, site 511\n",
      "\texchanged T for Y in sequence 23, site 1019\n",
      "\texchanged A for R in sequence 69, site 1039\n",
      "\texchanged A for R in sequence 88, site 1277\n",
      "\texchanged T for Y in sequence 80, site 1770\n",
      "\texchanged G for R in sequence 65, site 2852\n",
      "\texchanged A for R in sequence 81, site 3214\n",
      "\texchanged A for R in sequence 90, site 3885\n",
      "\texchanged A for R in sequence 2, site 4211\n",
      "\texchanged G for R in sequence 63, site 4228\n",
      "\texchanged C for Y in sequence 18, site 4425\n",
      "\texchanged G for R in sequence 63, site 4546\n",
      "703010159-5\n",
      "\tselected 108 of 247 sequences with <200 gaps in excess of consensus\n",
      "\tremoved 2 of 4256 sites with >95% gaps\n",
      "\texchanged C for N in sequence 96, site 236\n",
      "\texchanged A for N in sequence 96, site 237\n",
      "\texchanged A for R in sequence 76, site 383\n",
      "\texchanged A for R in sequence 80, site 545\n",
      "\texchanged A for R in sequence 63, site 720\n",
      "\texchanged A for R in sequence 39, site 909\n",
      "\texchanged C for Y in sequence 93, site 979\n",
      "\texchanged A for R in sequence 89, site 1016\n",
      "\texchanged A for R in sequence 77, site 1044\n",
      "\texchanged A for R in sequence 36, site 1054\n",
      "\texchanged A for N in sequence 6, site 1064\n",
      "\texchanged A for R in sequence 28, site 1163\n",
      "\texchanged A for N in sequence 99, site 1183\n",
      "\texchanged A for R in sequence 17, site 1648\n",
      "\texchanged A for R in sequence 53, site 1781\n",
      "\texchanged T for K in sequence 28, site 1872\n",
      "\texchanged A for R in sequence 82, site 2364\n",
      "\texchanged A for R in sequence 59, site 2380\n",
      "\texchanged C for Y in sequence 10, site 2498\n",
      "\texchanged A for R in sequence 97, site 2525\n",
      "\texchanged A for R in sequence 20, site 2653\n",
      "\texchanged A for R in sequence 29, site 2698\n",
      "\texchanged T for Y in sequence 66, site 2808\n",
      "\texchanged T for Y in sequence 83, site 2886\n",
      "\texchanged C for Y in sequence 20, site 2890\n",
      "\texchanged T for Y in sequence 84, site 3022\n",
      "\texchanged T for Y in sequence 63, site 3421\n",
      "\texchanged T for Y in sequence 84, site 3537\n",
      "\texchanged C for Y in sequence 60, site 3709\n",
      "\texchanged C for Y in sequence 60, site 3808\n",
      "\tmutant at site 4253 in codon that does not terminate in alignment, assuming syn\n",
      "\tmutant at site 4253 in codon that does not terminate in alignment, assuming syn\n",
      "703010159-3\n",
      "\tselected 140 of 247 sequences with <200 gaps in excess of consensus\n",
      "\tremoved 116 of 4744 sites with >95% gaps\n",
      "\texchanged A for R in sequence 42, site 57\n",
      "\texchanged A for R in sequence 108, site 181\n",
      "\texchanged T for Y in sequence 102, site 245\n",
      "\texchanged T for Y in sequence 76, site 299\n",
      "\texchanged A for R in sequence 40, site 360\n",
      "\texchanged G for N in sequence 53, site 689\n",
      "\texchanged A for R in sequence 137, site 983\n",
      "\texchanged C for N in sequence 53, site 1019\n",
      "\texchanged G for R in sequence 63, site 1142\n",
      "\texchanged T for W in sequence 8, site 1529\n",
      "\texchanged T for Y in sequence 6, site 1751\n",
      "\texchanged T for Y in sequence 126, site 1845\n",
      "\texchanged A for R in sequence 74, site 2287\n",
      "\texchanged C for Y in sequence 16, site 2340\n",
      "\texchanged A for R in sequence 16, site 2472\n",
      "\texchanged G for R in sequence 40, site 2615\n",
      "\texchanged T for Y in sequence 40, site 2663\n",
      "\texchanged T for N in sequence 53, site 2663\n",
      "\texchanged G for R in sequence 28, site 2666\n",
      "\texchanged T for Y in sequence 66, site 2738\n",
      "\texchanged A for R in sequence 64, site 3224\n",
      "\texchanged A for N in sequence 114, site 3311\n",
      "\texchanged G for R in sequence 107, site 3427\n",
      "\texchanged C for N in sequence 53, site 3528\n",
      "\texchanged C for Y in sequence 59, site 3533\n",
      "\texchanged G for R in sequence 3, site 4012\n",
      "\texchanged G for R in sequence 8, site 4012\n",
      "\texchanged G for R in sequence 16, site 4012\n",
      "\texchanged G for R in sequence 3, site 4015\n",
      "\texchanged A for N in sequence 53, site 4016\n",
      "\texchanged G for N in sequence 53, site 4017\n",
      "\texchanged T for N in sequence 53, site 4019\n",
      "\texchanged C for M in sequence 8, site 4036\n",
      "\texchanged C for M in sequence 16, site 4036\n",
      "\texchanged T for N in sequence 23, site 4480\n",
      "\tmutant at site 0 in codon that does not terminate in alignment, assuming syn\n",
      "\tmutant at site 0 in codon that does not terminate in alignment, assuming syn\n",
      "\tmutant at site 1 in codon that does not terminate in alignment, assuming syn\n",
      "\tmutant at site 1 in codon that does not terminate in alignment, assuming syn\n",
      "703010131-5\n",
      "\tselected 82 of 224 sequences with <200 gaps in excess of consensus\n",
      "\tremoved 2 of 4256 sites with >95% gaps\n",
      "\texchanged G for N in sequence 50, site 102\n",
      "\texchanged A for R in sequence 41, site 124\n",
      "\texchanged A for N in sequence 50, site 127\n",
      "\texchanged T for Y in sequence 24, site 148\n",
      "\texchanged A for R in sequence 44, site 313\n",
      "\texchanged A for R in sequence 22, site 853\n",
      "\texchanged G for N in sequence 50, site 1162\n",
      "\texchanged A for R in sequence 53, site 1376\n",
      "\texchanged T for Y in sequence 21, site 2146\n",
      "\texchanged C for N in sequence 50, site 2289\n",
      "\texchanged A for N in sequence 50, site 2395\n",
      "\texchanged A for R in sequence 13, site 2511\n",
      "\texchanged A for R in sequence 42, site 2973\n",
      "\texchanged A for R in sequence 60, site 3149\n",
      "\texchanged G for N in sequence 50, site 3211\n",
      "\texchanged A for N in sequence 50, site 3411\n",
      "\texchanged A for R in sequence 45, site 3724\n",
      "\texchanged G for R in sequence 52, site 4009\n",
      "\texchanged T for Y in sequence 44, site 4145\n",
      "\texchanged C for N in sequence 32, site 4168\n",
      "\tmutant at site 4253 in codon that does not terminate in alignment, assuming syn\n",
      "\tmutant at site 4253 in codon that does not terminate in alignment, assuming syn\n",
      "703010131-3\n",
      "\tselected 142 of 224 sequences with <200 gaps in excess of consensus\n",
      "\tremoved 80 of 4711 sites with >95% gaps\n",
      "\texchanged C for Y in sequence 90, site 145\n",
      "\texchanged T for Y in sequence 32, site 708\n",
      "\texchanged T for Y in sequence 136, site 708\n",
      "\texchanged T for Y in sequence 44, site 995\n",
      "\texchanged T for N in sequence 119, site 1041\n",
      "\texchanged T for W in sequence 90, site 1385\n",
      "\texchanged T for K in sequence 122, site 1385\n",
      "\texchanged G for R in sequence 13, site 1580\n",
      "\texchanged C for Y in sequence 90, site 1626\n",
      "\texchanged C for M in sequence 47, site 1826\n",
      "\texchanged C for Y in sequence 102, site 2231\n",
      "\texchanged T for N in sequence 75, site 2237\n",
      "\texchanged A for R in sequence 98, site 2271\n",
      "\texchanged A for R in sequence 4, site 2285\n",
      "\texchanged T for Y in sequence 122, site 2419\n",
      "\texchanged A for W in sequence 30, site 2478\n",
      "\texchanged A for R in sequence 96, site 2528\n",
      "\texchanged T for K in sequence 47, site 2618\n",
      "\texchanged G for N in sequence 131, site 2856\n",
      "\texchanged A for N in sequence 119, site 2912\n",
      "\texchanged A for R in sequence 2, site 3158\n",
      "\texchanged T for Y in sequence 60, site 3520\n",
      "\texchanged T for N in sequence 116, site 3855\n",
      "\texchanged A for R in sequence 47, site 4016\n",
      "\texchanged A for R in sequence 2, site 4178\n",
      "\texchanged A for R in sequence 56, site 4598\n",
      "\tmutant at site 0 in codon that does not terminate in alignment, assuming syn\n",
      "\tmutant at site 0 in codon that does not terminate in alignment, assuming syn\n",
      "700010607-5\n",
      "\tselected 76 of 141 sequences with <200 gaps in excess of consensus\n",
      "\tremoved 2 of 4256 sites with >95% gaps\n",
      "\tmutant at site 4253 in codon that does not terminate in alignment, assuming syn\n",
      "\tmutant at site 4253 in codon that does not terminate in alignment, assuming syn\n",
      "700010607-3\n",
      "\tselected 73 of 141 sequences with <200 gaps in excess of consensus\n",
      "\tremoved 59 of 4668 sites with >95% gaps\n",
      "\texchanged T for N in sequence 6, site 2546\n",
      "\texchanged T for N in sequence 3, site 3282\n",
      "700010470-5\n",
      "\tselected 104 of 217 sequences with <200 gaps in excess of consensus\n",
      "\tremoved 2 of 4286 sites with >95% gaps\n",
      "700010470-3\n",
      "\tselected 113 of 217 sequences with <200 gaps in excess of consensus\n",
      "\tremoved 47 of 4681 sites with >95% gaps\n",
      "700010077-5\n",
      "\tselected 43 of 156 sequences with <200 gaps in excess of consensus\n",
      "\tremoved 1 of 4259 sites with >95% gaps\n",
      "700010077-3\n",
      "\tselected 49 of 156 sequences with <200 gaps in excess of consensus\n",
      "\tremoved 42 of 4664 sites with >95% gaps\n",
      "\texchanged A for R in sequence 17, site 2244\n",
      "700010058-5\n",
      "\tselected 52 of 200 sequences with <200 gaps in excess of consensus\n",
      "\tremoved 4 of 4271 sites with >95% gaps\n",
      "\texchanged A for R in sequence 10, site 242\n",
      "\texchanged G for R in sequence 53, site 351\n",
      "\texchanged A for W in sequence 53, site 447\n",
      "\texchanged C for Y in sequence 27, site 718\n",
      "\texchanged A for R in sequence 10, site 1127\n",
      "\texchanged T for Y in sequence 41, site 1184\n",
      "\texchanged T for Y in sequence 44, site 1184\n",
      "\texchanged C for Y in sequence 29, site 1206\n",
      "\texchanged G for R in sequence 53, site 1531\n",
      "\texchanged A for R in sequence 5, site 1903\n",
      "\texchanged T for Y in sequence 13, site 2020\n",
      "\texchanged T for W in sequence 53, site 2113\n",
      "\texchanged G for R in sequence 13, site 2351\n",
      "\texchanged C for Y in sequence 8, site 2705\n",
      "700010058-3\n",
      "\tselected 28 of 200 sequences with <200 gaps in excess of consensus\n",
      "\tremoved 56 of 4649 sites with >95% gaps\n",
      "\texchanged A for R in sequence 3, site 205\n",
      "\texchanged A for R in sequence 3, site 2762\n",
      "\texchanged A for R in sequence 3, site 3083\n",
      "700010040-5\n",
      "\tselected 130 of 394 sequences with <200 gaps in excess of consensus\n",
      "\tremoved 13 of 4262 sites with >95% gaps\n",
      "\texchanged A for R in sequence 22, site 2273\n",
      "\texchanged A for R in sequence 5, site 2866\n",
      "\texchanged A for M in sequence 20, site 3275\n",
      "\tmutant at site 4247 in codon that does not terminate in alignment, assuming syn\n",
      "\tmutant at site 4247 in codon that does not terminate in alignment, assuming syn\n",
      "\tmutant at site 4248 in codon that does not terminate in alignment, assuming syn\n",
      "\tmutant at site 4248 in codon that does not terminate in alignment, assuming syn\n",
      "700010040-3\n",
      "\tselected 136 of 394 sequences with <200 gaps in excess of consensus\n",
      "\tremoved 62 of 4694 sites with >95% gaps\n",
      "\texchanged C for Y in sequence 11, site 1904\n",
      "\texchanged T for K in sequence 11, site 1926\n",
      "\texchanged A for R in sequence 15, site 2059\n",
      "\texchanged T for K in sequence 17, site 3938\n",
      "\texchanged T for Y in sequence 64, site 4005\n",
      "\texchanged A for R in sequence 16, site 4213\n",
      "\texchanged G for K in sequence 30, site 4231\n",
      "\texchanged C for M in sequence 17, site 4401\n",
      "\texchanged G for R in sequence 32, site 4412\n",
      "\texchanged G for R in sequence 32, site 4424\n",
      "cap256-3\n",
      "\tselected 161 of 176 sequences with <200 gaps in excess of consensus\n",
      "\tremoved 5 of 2651 sites with >95% gaps\n",
      "\texchanged C for Y in sequence 150, site 162\n",
      "\texchanged G for N in sequence 111, site 326\n",
      "\texchanged T for Y in sequence 19, site 329\n",
      "\texchanged C for N in sequence 111, site 339\n",
      "\texchanged A for N in sequence 111, site 380\n",
      "\texchanged A for N in sequence 110, site 650\n",
      "\texchanged T for N in sequence 111, site 734\n",
      "\texchanged G for N in sequence 111, site 1086\n",
      "\texchanged A for R in sequence 19, site 1236\n",
      "\texchanged T for N in sequence 111, site 1245\n",
      "\texchanged G for R in sequence 111, site 1428\n",
      "\texchanged G for R in sequence 19, site 1573\n",
      "\texchanged C for Y in sequence 19, site 2223\n",
      "\texchanged G for N in sequence 115, site 2244\n",
      "\texchanged A for R in sequence 19, site 2558\n",
      "\tmutant at site 0 in codon that does not terminate in alignment, assuming syn\n",
      "\tmutant at site 0 in codon that does not terminate in alignment, assuming syn\n",
      "\tmutant at site 1 in codon that does not terminate in alignment, assuming syn\n",
      "\tmutant at site 1 in codon that does not terminate in alignment, assuming syn\n",
      "\n",
      "Out of 350045 possible mutations,\n",
      "\t6803 (1.94%) nonsynonymous in epitope\n",
      "\t4383 (1.25%) nonsynonymous reversion\n",
      "\t118 (0.03%) nonsynonymous reversion, in epitope\n",
      "\t4265 (1.22%) nonsynonymous reversion, nonepitope\n",
      "\t74253 (21.21%) synonymous\n"
     ]
    }
   ],
   "source": [
    "# Load the sequence coverage for each patient\n",
    "df_range = pd.read_csv('%s/range.csv' % (HIV_DIR), comment='#', memory_map=True)\n",
    "\n",
    "# Load the list of transmitted/founder (TF) sequences\n",
    "df_TF = pd.read_csv('%s/TF.csv' % (HIV_DIR), comment='#', memory_map=True)\n",
    "\n",
    "# Load the list of exposed residues on Env\n",
    "df_exposed = pd.read_csv('%s/exposed.csv' % (HIV_DIR), comment='#', memory_map=True)\n",
    "\n",
    "# Load the set of epitopes targeted by patients\n",
    "df_epitope = pd.read_csv('%s/epitopes.csv' % (HIV_DIR), comment='#', dtype={'ppt': str}, memory_map=True)\n",
    "\n",
    "\n",
    "# Association times with sequences, remove all sequences (except HXB2 and consensus)\n",
    "# that do not have identifiable times; TF sequence is treated as day 0 if available\n",
    "for ppt in set(df_range.ppt):\n",
    "    HIV.filter_sequences(ppt, '%s/raw/%s-HIValign.fasta' % (HIV_DIR, ppt), check_time=True)\n",
    "    \n",
    "    \n",
    "# Store baseline number of nonsynonymous mutations in CD8+ T cell epitopes, \n",
    "# nonsynonymous reversions (total, in epitopes, and outside epitopes), and synonymous mutations\n",
    "BASELINE_NS_EPITOPE = 0\n",
    "BASELINE_NS_REV = 0\n",
    "BASELINE_NS_REV_NONEPI = 0\n",
    "BASELINE_SYN = 0\n",
    "BASELINE_TOT = 0\n",
    "\n",
    "# Select and process alignment subsets for analysis\n",
    "g = open('%s/run.sh' % (MPL_DIR), 'w')\n",
    "g.write('g++ src/main.cpp src/inf.cpp src/io.cpp -O3 -march=native -lgslcblas -lgsl -o bin/mpl\\n')\n",
    "\n",
    "for df_range_iter, df_range_entry in df_range.iterrows():\n",
    "    \n",
    "    # Read in MSA\n",
    "    ppt = df_range_entry.ppt\n",
    "    msa, tag = HIV.get_MSA('%s/interim/%s-HIValign-filtered.fasta' % (HIV_DIR, ppt), noArrow=True)\n",
    "    print('%s' % df_range_entry.tag)\n",
    "    \n",
    "    # Clip MSA to specified range\n",
    "    msa = HIV.clip_MSA(df_range_entry.start, df_range_entry.end, msa, tag)\n",
    "    \n",
    "    # Filter sequences for quality (maximum # excess gaps)\n",
    "    max_gap_num = 200\n",
    "    max_gap_freq = 0.05\n",
    "    current_msa, current_tag = HIV.filter_excess_gaps(msa, tag, max_gap_num, max_gap_freq, verbose=True)\n",
    "    \n",
    "    # Put sequences in time order (first entries are HXB2, consensus)\n",
    "    current_msa, current_tag = HIV.order_sequences(current_msa, current_tag)\n",
    "    \n",
    "    # Locate or compute TF sequence\n",
    "    TF_accession = df_TF[df_TF.ppt==ppt].iloc[0].TF\n",
    "    TF_sequence = HIV.get_TF(current_msa, current_tag, TF_accession)\n",
    "    \n",
    "    temp_msa = [current_msa[list(current_tag).index(HXB2_TAG)], TF_sequence]\n",
    "    temp_tag = [HXB2_TAG, '%s-TF' % ppt]\n",
    "    HIV.save_MSA(temp_msa, temp_tag, '%s/interim/TF-%s' % (HIV_DIR, df_range_entry.tag))\n",
    "    \n",
    "    # Impute ambiguous nucleotides and gaps at start/end of alignment\n",
    "    current_msa = HIV.impute_ambiguous(current_msa, current_tag, start_index=2, verbose=True)\n",
    "    \n",
    "    # Map between local (alignment) index, index of polymorphic sites, and HXB2\n",
    "    # Track location of known T cell epitopes, TF sequence, and exposed residues of Env\n",
    "    min_seqs = 4\n",
    "    max_dt = 300\n",
    "    poly_sites, current_msa, current_tag = HIV.create_index(current_msa, current_tag, TF_sequence, \n",
    "                                           current_msa[list(current_tag).index(CONS_TAG)], \n",
    "                                           current_msa[list(current_tag).index(HXB2_TAG)],\n",
    "                                           df_range_entry.start, min_seqs, max_dt, \n",
    "                                           df_epitope[df_epitope.ppt==ppt], df_exposed, \n",
    "                                           '%s/processed/%s-index.csv' % (HIV_DIR, df_range_entry.tag), \n",
    "                                           return_polymorphic=True, return_truncated=True)\n",
    "    \n",
    "    # Save alignment to format readable by MPL\n",
    "    poly_states, poly_times = HIV.save_MPL_alignment(current_msa, current_tag, \n",
    "                                                     '%s/HIV/%s-poly-seq2state.dat' % (MPL_DIR, df_range_entry.tag), \n",
    "                                                     polymorphic_sites=poly_sites, return_states=True)\n",
    "    \n",
    "    # Save allele frequency trajectories for polymorphic sites\n",
    "    # Later this file will be edited to include inferred selection coefficients\n",
    "    df_index = pd.read_csv('%s/processed/%s-index.csv' % (HIV_DIR, df_range_entry.tag), comment='#', memory_map=True)\n",
    "    HIV.save_trajectories(poly_sites, poly_states, poly_times, TF_sequence, df_index,\n",
    "                          '%s/interim/%s-poly.csv' % (HIV_DIR, df_range_entry.tag))\n",
    "    \n",
    "    # Get baseline number of possible mutations in different categories\n",
    "    if ppt not in ['cap256']:\n",
    "        df_poly = pd.read_csv('%s/interim/%s-poly.csv' % (HIV_DIR, df_range_entry.tag), comment='#', memory_map=True)\n",
    "        temp_ns_epitope, temp_ns_rev, temp_ns_rev_ne, temp_syn = HIV.get_baseline(TF_sequence, df_index, df_poly)\n",
    "        BASELINE_NS_EPITOPE += temp_ns_epitope\n",
    "        BASELINE_NS_REV += temp_ns_rev\n",
    "        BASELINE_NS_REV_NONEPI += temp_ns_rev_ne\n",
    "        BASELINE_SYN += temp_syn\n",
    "        BASELINE_TOT += 3 * len(df_index) + np.sum(df_poly.nucleotide=='-')\n",
    "    \n",
    "    # Save MSA\n",
    "    HIV.save_MSA(current_msa, current_tag, '%s/interim/%s' % (HIV_DIR, df_range_entry.tag))\n",
    "    \n",
    "    # Record instructions for running MPL on this data\n",
    "    io_str = '%s-poly-seq2state' % df_range_entry.tag\n",
    "    run_str = ('./bin/mpl -d HIV -i %s.dat -o %s-MPL.dat -g 5e4 -N 1e4 -m Zanini-extended.dat -sc %s -sn %s\\n'+\n",
    "               './bin/mpl -d HIV -i %s.dat -o %s-SL.dat  -g 5e4 -N 1e4 -m Zanini-extended.dat -nc\\n')\n",
    "    g.write(run_str % (io_str, io_str, 'covariance-'+io_str+'.dat', 'numerator-'+io_str+'.dat', io_str, io_str))\n",
    "\n",
    "g.close()\n",
    "\n",
    "\n",
    "# Print baseline numbers of possible mutations in different categories\n",
    "print('\\nOut of %d possible mutations,' % BASELINE_TOT)\n",
    "print('\\t%d (%.2f%%) nonsynonymous in epitope' \n",
    "      % (BASELINE_NS_EPITOPE, 100*BASELINE_NS_EPITOPE/BASELINE_TOT))\n",
    "print('\\t%d (%.2f%%) nonsynonymous reversion'\n",
    "      % (BASELINE_NS_REV, 100*BASELINE_NS_REV/BASELINE_TOT))\n",
    "print('\\t%d (%.2f%%) nonsynonymous reversion, in epitope' \n",
    "      % (BASELINE_NS_REV-BASELINE_NS_REV_NONEPI, 100*(BASELINE_NS_REV-BASELINE_NS_REV_NONEPI)/BASELINE_TOT))\n",
    "print('\\t%d (%.2f%%) nonsynonymous reversion, nonepitope' \n",
    "      % (BASELINE_NS_REV_NONEPI, 100*BASELINE_NS_REV_NONEPI/BASELINE_TOT))\n",
    "print('\\t%d (%.2f%%) synonymous'\n",
    "      % (BASELINE_SYN, 100*BASELINE_SYN/BASELINE_TOT))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Add superinfection sequence (SU) to index CSV for CAP256"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\tselected 168 of 183 sequences with <200 gaps in excess of consensus\n",
      "\tremoved 5 of 2651 sites with >95% gaps\n"
     ]
    }
   ],
   "source": [
    "# Load the sequence coverage for each patient\n",
    "df_range = pd.read_csv('%s/range.csv' % (HIV_DIR), comment='#', memory_map=True)\n",
    "\n",
    "# Read in MSA\n",
    "ppt = 'cap256'\n",
    "HIV.filter_sequences(ppt+'-SU', '%s/raw/%s-HIValign.fasta' % (HIV_DIR, ppt+'-SU'), check_time=True)\n",
    "msa, tag = HIV.get_MSA('%s/interim/%s-HIValign-filtered.fasta' % (HIV_DIR, ppt+'-SU'), noArrow=True)\n",
    "\n",
    "# Clip MSA to specified range\n",
    "df_range = df_range[df_range.ppt=='cap256']\n",
    "msa = HIV.clip_MSA(df_range.iloc[0].start, df_range.iloc[0].end, msa, tag)\n",
    "\n",
    "# Filter sequences for quality (maximum # excess gaps)\n",
    "max_gap_num = 200\n",
    "max_gap_freq = 0.05\n",
    "current_msa, current_tag = HIV.filter_excess_gaps(msa, tag, max_gap_num, max_gap_freq, verbose=True)\n",
    "    \n",
    "# Put sequences in time order (first entries are HXB2, consensus)\n",
    "current_msa, current_tag = HIV.order_sequences(current_msa, current_tag)\n",
    "    \n",
    "# Compute SU sequence\n",
    "SU_accession = 'avg'\n",
    "SU_sequence  = HIV.get_TF(current_msa, current_tag, SU_accession)\n",
    "\n",
    "# Add SU sequence to index CSV\n",
    "df_index = pd.read_csv('%s/processed/%s-3-index.csv' % (HIV_DIR, ppt), comment='#', memory_map=True)\n",
    "df_index = df_index.assign(SU=pd.Series(list(SU_sequence)))\n",
    "df_index.to_csv('%s/processed/%s-3-index.csv' % (HIV_DIR, ppt+'-SU'), index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Shift selection coefficients\n",
    "\n",
    "After selection has been inferred, the selection coefficients should be normalized such that the TF nucleotide at each site has a selection coefficient of zero. The cell below performs this normalization and saves all the selection coefficients together in the file `HIV/analysis/total-selection.csv`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Shift selection coefficients\n",
    "\n",
    "df_range = pd.read_csv('%s/range.csv' % (HIV_DIR), comment='#', memory_map=True)\n",
    "df_epitope = pd.read_csv('%s/epitopes.csv' % (HIV_DIR), comment='#', dtype={'ppt': str}, memory_map=True)\n",
    "\n",
    "f = open('%s/analysis/total-selection.csv' % (HIV_DIR), 'w')\n",
    "f.write('ppt,tag,polymorphic_index,alignment_index,HXB2_index,consensus_nucleotide,TF_nucleotide,nucleotide,' +\n",
    "        'in_epitope,epitope,epitope_detected,flanking,exposed,glycan,nonsynonymous,s_MPL,s_SL\\n')\n",
    "\n",
    "# potential future comparisons: set point, clade, time epitope targeted, immunodominance, M/F, age\n",
    "\n",
    "for idx, entry in df_range.iterrows():\n",
    "    ppt = entry.ppt\n",
    "    \n",
    "    # Read in MPL output\n",
    "    temp_MPL = np.loadtxt('%s/HIV/%s-poly-seq2state-MPL.dat' % (MPL_DIR, entry.tag))\n",
    "    temp_SL = np.loadtxt('%s/HIV/%s-poly-seq2state-SL.dat' % (MPL_DIR, entry.tag))\n",
    "    L = int(len(temp_MPL)/len(NUC))\n",
    "    temp_MPL = np.reshape(temp_MPL, (L, len(NUC)))\n",
    "    temp_SL = np.reshape(temp_SL, (L, len(NUC)))\n",
    "    \n",
    "    df_info = pd.read_csv('%s/interim/%s-poly.csv' % (HIV_DIR, entry.tag), comment='#', memory_map=True)\n",
    "    \n",
    "    # Normalize selection coefficients\n",
    "    s_MPL = []\n",
    "    s_SL = []\n",
    "    for i in range(L):\n",
    "        TF = df_info[df_info.polymorphic_index==i].iloc[0].TF\n",
    "        idx = NUC.index(TF)\n",
    "        \n",
    "        temp_s = temp_MPL[i]\n",
    "        temp_s = temp_s - temp_s[idx]\n",
    "        s_MPL.append(temp_s)\n",
    "    \n",
    "        temp_s = temp_SL[i]\n",
    "        temp_s = temp_s - temp_s[idx]\n",
    "        s_SL.append(temp_s)\n",
    "    \n",
    "    reduced_s_MPL = []\n",
    "    reduced_s_SL = []\n",
    "    for df_info_iter, df_info_entry in df_info.iterrows():\n",
    "        i = df_info_entry.polymorphic_index\n",
    "        j = NUC.index(df_info_entry.nucleotide)\n",
    "        reduced_s_MPL.append(s_MPL[i][j])\n",
    "        reduced_s_SL.append(s_SL[i][j])\n",
    "    \n",
    "    # Record data\n",
    "    df_info['s_MPL'] = pd.Series(reduced_s_MPL, index=df_info.index)\n",
    "    df_info['s_SL'] = pd.Series(reduced_s_SL, index=df_info.index)\n",
    "    \n",
    "    df_info.to_csv('%s/analysis/%s-poly.csv' % (HIV_DIR, entry.tag), index=False)\n",
    "    \n",
    "    mean_MPL = np.mean(df_info.s_MPL)\n",
    "    std_MPL = np.std(df_info.s_MPL)\n",
    "    mean_SL = np.mean(df_info.s_SL)\n",
    "    std_SL = np.std(df_info.s_SL)\n",
    "    \n",
    "    for df_info_iter, df_info_entry in df_info.iterrows():\n",
    "        if df_info_entry.TF==df_info_entry.nucleotide:\n",
    "            continue\n",
    "        detected = 'nan'\n",
    "        if pd.notnull(df_info_entry.epitope):\n",
    "            detected = str(int(df_epitope[((df_epitope.ppt==ppt) & \n",
    "                                           (df_epitope.epitope==df_info_entry.epitope))].detected))\n",
    "        f.write('%s,%s,%s,%d,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%.4f,%.4f\\n' % \n",
    "                (ppt, entry.tag, df_info_entry.polymorphic_index, df_info_entry.alignment_index, \n",
    "                 df_info_entry.HXB2_index, df_info_entry.consensus, df_info_entry.TF,\n",
    "                 df_info_entry.nucleotide, not pd.isnull(df_info_entry.epitope), df_info_entry.epitope, \n",
    "                 detected, str(int(df_info_entry.flanking)), df_info_entry.exposed,\n",
    "                 str(int(df_info_entry.glycan)), str(int(df_info_entry.nonsynonymous)), \n",
    "                 df_info_entry.s_MPL, df_info_entry.s_SL))\n",
    "\n",
    "f.close()    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Quantify influence of individual variants on inferred selection at other sites\n",
    "\n",
    "The selection coefficients that we infer depend not only on the trajectories of individual variants, but also on the genetic background(s) on which they arise. To quantify this, we generate *masked* trajectories where individual variants are replaced by the TF nucleotide. We can then compare selection inferred with and without masking to determine how each variant affects estimates of selection for other variants. In the cells below we generate and save the masked trajectories, then we normalize the and compare the inferred selection with and without masking."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate masked trajectories and associated job files for MPL\n",
    "\n",
    "f = open('%s/jobs/run-ds.sh' % (MPL_DIR), 'w')\n",
    "\n",
    "df_range = pd.read_csv('%s/range.csv' % (HIV_DIR), comment='#', memory_map=True)\n",
    "for it, entry in df_range.iterrows():\n",
    "    tag = entry.tag\n",
    "    df = pd.read_csv('%s/analysis/%s-poly.csv' % (HIV_DIR, tag), comment='#', memory_map=True)\n",
    "    df_noTF = df[df.nucleotide!=df.TF]\n",
    "\n",
    "    # Get sequence data\n",
    "    traj = np.loadtxt('%s/HIV/%s-poly-seq2state.dat' % (MPL_DIR, tag))\n",
    "    \n",
    "    for var_it, var_entry in df_noTF.iterrows():\n",
    "        var_id = str(int(var_entry.polymorphic_index))+var_entry.nucleotide\n",
    "        \n",
    "        # Save masked trajectory\n",
    "        temp_traj = np.copy(traj)\n",
    "        poly_idxs = traj.T[var_entry.polymorphic_index+2]==NUC.index(var_entry.nucleotide)\n",
    "        temp_traj.T[var_entry.polymorphic_index+2][poly_idxs] = NUC.index(var_entry.TF)\n",
    "        np.savetxt('%s/HIV/mask/%s-poly-seq2state-%s.dat' % (MPL_DIR, tag, var_id), temp_traj, fmt='%d')\n",
    "    \n",
    "        # Write job to file\n",
    "        f.write('./bin/mpl -d HIV/mask -i %s-poly-seq2state-%s.dat -o %s-poly-seq2state-MPL-%s.dat' \n",
    "                % (tag, var_id, tag, var_id))\n",
    "        f.write(' -g 5e4 -N 1e4 -m Zanini-extended.dat\\n')\n",
    "\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Shift and compare selection coefficients\n",
    "\n",
    "df_range = pd.read_csv('%s/range.csv' % (HIV_DIR), comment='#', memory_map=True)\n",
    "for it, entry in df_range.iterrows():\n",
    "    \n",
    "    # Read in information on polymorphic sites, TF, selection\n",
    "    tag = entry.tag\n",
    "    df = pd.read_csv('%s/analysis/%s-poly.csv' % (HIV_DIR, tag), comment='#', memory_map=True)\n",
    "    L = len(np.unique(df.polymorphic_index))\n",
    "    TF_idxs = [NUC.index(df[df.polymorphic_index==i].iloc[0].TF) for i in range(L)]\n",
    "    \n",
    "    f = open('%s/analysis/%s-delta-s.csv' % (HIV_DIR, tag), 'w')\n",
    "    f.write('mask_polymorphic_index,mask_nucleotide,target_polymorphic_index,target_nucleotide,effect,distance\\n')\n",
    "    \n",
    "    # Iterate through masked variants\n",
    "    df_noTF = df[df.nucleotide!=df.TF]\n",
    "    for var_it, var_entry in df_noTF.iterrows():\n",
    "        var_id = str(int(var_entry.polymorphic_index))+var_entry.nucleotide\n",
    "        \n",
    "        temp_MPL = np.loadtxt('%s/HIV/mask/%s-poly-seq2state-MPL-%s.dat' % (MPL_DIR, entry.tag, var_id))\n",
    "        temp_MPL = np.reshape(temp_MPL, (L, len(NUC)))\n",
    "        mask_MPL = [temp_MPL[i] - temp_MPL[i][TF_idxs[i]] for i in range(L)]\n",
    "        \n",
    "        for target_it, target_entry in df_noTF.iterrows():\n",
    "            f.write('%d,%s,%d,%s,' % (int(var_entry.polymorphic_index), var_entry.nucleotide, \n",
    "                                      int(target_entry.polymorphic_index), target_entry.nucleotide))\n",
    "            f.write('%le,' % (target_entry.s_MPL \n",
    "                              - mask_MPL[int(target_entry.polymorphic_index)][NUC.index(target_entry.nucleotide)]))\n",
    "            f.write('%d\\n' % (np.fabs(int(var_entry.alignment_index) - int(target_entry.alignment_index))))\n",
    "    \n",
    "    f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
